{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e4387d2-7683-4c42-93b4-5776e7107cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import h5py \n",
    "import os\n",
    "\n",
    "from cerberus.cerberus import *\n",
    "from cerberus.main import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0faa232-0c52-4e30-9591-96406c346324",
   "metadata": {},
   "source": [
    "## modify / collapse transcripts in TALON GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c0b46624-a1a1-43ca-b1dd-0262ef3c1180",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_stranded_gtf_dfs(df):\n",
    "    \"\"\"\n",
    "    Split a GTF df into fwd and rev strands\n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): DF of gtf\n",
    "    \n",
    "    Returns:\n",
    "        fwd (pandas DataFrame): DF of all forward-stranded entries from GTF\n",
    "        rev (pandas DataFrame): DF of all reverse-stranded entries from GTF\n",
    "    \"\"\"\n",
    "    rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "    fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "    \n",
    "    return fwd, rev\n",
    "\n",
    "def sort_gtf(df):\n",
    "    \"\"\"\n",
    "    Sort a GTF into its proper ordering\n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): DF of GTF\n",
    "    \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF of GTF, sorted\n",
    "    \"\"\"\n",
    "    df['feature_rank'] = df.Feature.map({'gene':0, 'transcript':1, 'exon':2})\n",
    "    df.feature_rank = df.feature_rank.astype(int)    \n",
    "     \n",
    "    fwd, rev = get_stranded_gtf_dfs(df)\n",
    "\n",
    "    df = pd.DataFrame()\n",
    "    for temp in [fwd, rev]:\n",
    "        if len(temp.index) > 0:\n",
    "            strand = temp.Strand.values.tolist()[0]\n",
    "            if strand == '+':\n",
    "                ascending = True\n",
    "            elif strand == '-':\n",
    "                ascending = False\n",
    "            temp.sort_values(by=['gene_id', 'transcript_id', 'feature_rank', 'Start'], \n",
    "                             ascending=[True, True, True, ascending],\n",
    "                             na_position='first', inplace=True)\n",
    "            \n",
    "            df = pd.concat([df, temp], ignore_index=True)\n",
    "    df.drop('feature_rank', axis=1, inplace=True)\n",
    "    return df\n",
    "\n",
    "def get_update_ends_settings(strand, mode):\n",
    "    \"\"\"\n",
    "    Returns which columns to refer to and which min/max function\n",
    "    to use depending on looking at forward / rev strand or \n",
    "    tss / tes\n",
    "    \n",
    "    Parameters:\n",
    "        strand (str): {'+', '-'}\n",
    "        mode (str): {'tss', 'tes'}\n",
    "    \n",
    "    Returns:\n",
    "        old_end (str): Name of column to modify; {'Start', 'End'}\n",
    "        new_end (str): Name of column to pull new value from; {'Start_end', 'End_end'}\n",
    "        gene_func (str): What function to apply to new_end; {'min', 'max'}\n",
    "    \"\"\"\n",
    "    if mode == 'tss': \n",
    "        if strand == '+':\n",
    "            old_end = 'Start'\n",
    "            new_end = 'Start_end'\n",
    "            gene_func = 'min'\n",
    "        elif strand == '-':\n",
    "            old_end = 'End'\n",
    "            new_end = 'End_end'\n",
    "            gene_func = 'max'\n",
    "    elif mode == 'tes':\n",
    "        if strand == '+':\n",
    "            old_end = 'End'\n",
    "            new_end = 'End_end'\n",
    "            gene_func = 'max'\n",
    "        elif strand == '-':\n",
    "            old_end = 'Start'\n",
    "            new_end  = 'Start_end'\n",
    "            gene_func = 'min'\n",
    "            \n",
    "    return old_end, new_end, gene_func\n",
    "\n",
    "def update_transcript_ends(df, mode, strand):\n",
    "    \"\"\"\n",
    "    Update the ends of transcripts and the first / last exon\n",
    "    in a GTF. GTF must be sorted!\n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): Sorted DF of GTF with 'Start_end', and 'End_end' \n",
    "            columns denoting the boundaries of each end region\n",
    "        mode (str): {'tss', 'tes'}\n",
    "        strand (str): {'+', '-'}\n",
    "    \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF of GTF with transcript and \n",
    "            exon ends modified\n",
    "    \"\"\"\n",
    "    old_col, new_col, gene_func = get_update_ends_settings(strand, mode)\n",
    "    \n",
    "    temp = df[['Feature', 'gene_id', 'transcript_id', 'Strand', 'Start', 'End', 'Start_end', 'End_end']].copy(deep=True)\n",
    "    temp = temp.loc[temp.Feature != 'gene']\n",
    "    if mode == 'tss':\n",
    "        inds = temp.groupby('transcript_id').head(2).index.tolist()\n",
    "    elif mode == 'tes':\n",
    "        inds = temp.groupby('transcript_id').head(1).index.tolist()\n",
    "        inds += temp.groupby('transcript_id').tail(1).index.tolist()\n",
    "\n",
    "    df.loc[inds, old_col] = df.loc[inds, new_col]\n",
    "    \n",
    "    # convert float dtypes\n",
    "    df.Start = df.Start.astype(int)\n",
    "    df.End = df.End.astype(int)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def update_gene_ends(df, mode, strand):\n",
    "    \"\"\"\n",
    "    Update the ends of genes in a GTF. \n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): GTF dataframe \n",
    "        mode (str): {'tss', 'tes'}\n",
    "        strand (str): {'+', '-'}\n",
    "        \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame of GTF with gene ends updated\n",
    "    \"\"\"\n",
    "    # determine which ends we're updating and how we're doing so \n",
    "    old_col, new_col, gene_func = get_update_ends_settings(strand, mode)\n",
    "    \n",
    "    # get min or max of transcript ends depending on settings\n",
    "    temp = df[['Feature', 'gene_id', old_col]].copy(deep=True)\n",
    "    temp = temp.loc[temp.Feature == 'transcript']\n",
    "    temp = temp.groupby(['gene_id', 'Feature'], observed=True).agg(gene_func).reset_index()\n",
    "    temp.drop('Feature', axis=1, inplace=True)\n",
    "    \n",
    "    # add that coord to gene end\n",
    "    df = df.merge(temp, on='gene_id', suffixes=('', '_gene'))\n",
    "    inds = df.loc[df.Feature == 'gene'].index.tolist()\n",
    "    df.loc[inds, old_col] = df.loc[inds, '{}_gene'.format(old_col)]\n",
    "    df.drop('{}_gene'.format(old_col), axis=1, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def update_gtf_ends(gtf, tss, tes):\n",
    "    \"\"\"\n",
    "    Update gene, transcript, and exon boundaries to be \n",
    "    furthest upstream or downstream entry for end used\n",
    "    \n",
    "    Parameters:\n",
    "        gtf (pandas DataFrame): DF of GTF\n",
    "        tss (pyranges PyRanges): PyRanges object of reference TSSs\n",
    "        tes (pyranges PyRanges): PyRanges object of reference TESs\n",
    "    \n",
    "    Returns: \n",
    "        gtf (pandas DataFrame): DF of GTF with updated ends\n",
    "            based on the TSSs and TESs used in the input beds\n",
    "    \"\"\"\n",
    "    gtf = gtf.copy(deep=True)\n",
    "\n",
    "    for mode, ends in zip(['tss', 'tes'], [tss, tes]):\n",
    "        ends = ends.df\n",
    "        ends = ends[['Start', 'End', '{}_id'.format(mode)]]\n",
    "        gtf = gtf.merge(ends, how='left',\n",
    "                        on='{}_id'.format(mode),\n",
    "                        suffixes=('', '_end'))\n",
    "\n",
    "        fwd, rev = get_stranded_gtf_dfs(gtf)\n",
    "        df = pd.DataFrame()\n",
    "        for strand, temp in zip(['+', '-'], [fwd, rev]):\n",
    "    \n",
    "            # fix exon, transcript, and gene boundaries\n",
    "            temp = update_transcript_ends(temp, mode, strand)\n",
    "            temp = update_gene_ends(temp, mode, strand)\n",
    "            df = pd.concat([df, temp], ignore_index=True)\n",
    "        \n",
    "        gtf = df.copy(deep=True)\n",
    "        gtf.drop(['Start_end', 'End_end'], axis=1, inplace=True)\n",
    "        \n",
    "    return gtf\n",
    "\n",
    "def agg_gtf(df):\n",
    "    \"\"\"\n",
    "    Deduplicate GTF transcripts that have the same triplet id\n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): DF of gtf from `update_ends` \n",
    "    \n",
    "    Returns:\n",
    "        df (pandas DataFrame): DF of gtf with deduplicated\n",
    "            transcript / exon entries based on the triplet id\n",
    "    \"\"\"\n",
    "    \n",
    "    def collapse_non_gb_col(x):\n",
    "        x = x.astype(str)\n",
    "        x = x.unique().tolist()\n",
    "        x = ','.join(x)\n",
    "        return(x)\n",
    "\n",
    "    gb_cols = ['Chromosome',\n",
    "                 'Feature', \n",
    "                 'Start', 'End', \n",
    "                 'Score', 'Strand', 'Frame', 'gene_id', 'gene_name', \n",
    "                 'gene_status', 'gene_type', 'talon_gene', \n",
    "                 'ic', 'ic_id', 'tss_id', 'tss', 'tes_id', 'tes', 'transcript_id',\n",
    "                 'transcript_name']\n",
    "    gb_cols = list(set(df.columns)&set(gb_cols))\n",
    "\n",
    "    # get collapsed features to add to deduplicated df\n",
    "    t_df = df.loc[df.Feature == 'transcript'].copy(deep=True)\n",
    "    non_gb_cols = list(set(t_df.columns.tolist())-set(gb_cols))\n",
    "    t_df[non_gb_cols].fillna('', inplace=True)\n",
    "    t_df = t_df.groupby(gb_cols, observed=True)[non_gb_cols].agg(collapse_non_gb_col).reset_index()\n",
    "    t_df = t_df[['transcript_id']+non_gb_cols]\n",
    "    collapsed_feats = t_df.copy(deep=True)\n",
    "\n",
    "    # deduplicate df based only on transcript id\n",
    "    temp = df[['transcript_id', 'original_transcript_id']].drop_duplicates()\n",
    "    dupe_old_tids = temp.loc[temp.transcript_id.duplicated(keep='first'), 'original_transcript_id']\n",
    "    df = df.loc[~df.original_transcript_id.isin(dupe_old_tids)]\n",
    "    \n",
    "    # replace the non gb columns with the ones that we already grouped\n",
    "    df.drop(non_gb_cols, axis=1, inplace=True)\n",
    "    df = df.merge(collapsed_feats, how='left', on='transcript_id')\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "7db16831-2564-4356-ba8d-799ef02cde42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3941882\n",
      "     transcript_id_cerberus        transcript_id      Start  Start_end\n",
      "12  ENSG00000000460[3,10,3]    ENCODEHT001085918  169794729  169794679\n",
      "37  ENSG00000000460[3,10,3]  ENCODEHT001085918#0  169794782  169794679\n",
      "62  ENSG00000000460[3,10,3]  ENCODEHT001085918#1  169794782  169794679\n",
      "     transcript_id_cerberus        transcript_id      Start  Start_end\n",
      "12  ENSG00000000460[3,10,3]    ENCODEHT001085918  169794679  169794679\n",
      "37  ENSG00000000460[3,10,3]  ENCODEHT001085918#0  169794679  169794679\n",
      "62  ENSG00000000460[3,10,3]  ENCODEHT001085918#1  169794679  169794679\n",
      "Empty DataFrame\n",
      "Columns: [transcript_id_cerberus, transcript_id, End, End_end]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [transcript_id_cerberus, transcript_id, End, End_end]\n",
      "Index: []\n",
      "     transcript_id_cerberus        transcript_id        End    End_end\n",
      "12  ENSG00000000460[3,10,3]    ENCODEHT001085918  169853085  169853135\n",
      "37  ENSG00000000460[3,10,3]  ENCODEHT001085918#0  169852909  169853135\n",
      "62  ENSG00000000460[3,10,3]  ENCODEHT001085918#1  169853085  169853135\n",
      "     transcript_id_cerberus        transcript_id        End    End_end\n",
      "12  ENSG00000000460[3,10,3]    ENCODEHT001085918  169853135  169853135\n",
      "37  ENSG00000000460[3,10,3]  ENCODEHT001085918#0  169853135  169853135\n",
      "62  ENSG00000000460[3,10,3]  ENCODEHT001085918#1  169853135  169853135\n",
      "Empty DataFrame\n",
      "Columns: [transcript_id_cerberus, transcript_id, Start, Start_end]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [transcript_id_cerberus, transcript_id, Start, Start_end]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "h5 = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/cerberus/human_cerberus.h5'\n",
    "# gtf = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/talon/human_known_nic_nnc_talon.gtf'\n",
    "gtf = '/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/lapa/human_swan_talon.corrected.gtf'\n",
    "# gtf = 'test_dupe_genes.gtf'\n",
    "# gtf = 'test_c1orf112.gtf'\n",
    "agg = True\n",
    "update_ends = True\n",
    "\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b2c13b87-5375-489e-9364-f9bc08cb9e7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_back = df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81ee6383-c1ad-43bc-8372-2c10a619ee95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # get test set to work with \n",
    "# df = df_back.copy(deep=True)\n",
    "# temp = df.loc[df.Feature == 'transcript'].copy(deep=True)\n",
    "# temp = temp.loc[temp.transcript_id.duplicated(keep=False)]\n",
    "# temp.head()\n",
    "# gids_1 = temp.loc[temp.Strand == '+', 'gene_id'].unique().tolist()[:5]\n",
    "# gids_2 = temp.loc[temp.Strand == '-', 'gene_id'].unique().tolist()[:5]\n",
    "# gids = gids_1+gids_2\n",
    "\n",
    "# # # temp.head() \n",
    "# # temp.gene_id.head()\n",
    "# gids\n",
    "\n",
    "# df_back = df.copy(deep=True)\n",
    "\n",
    "# df = df.loc[df.gene_id.isin(gids)]\n",
    "# df = pr.PyRanges(df)\n",
    "# df.to_gtf('test_dupe_genes.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3079607b-9963-4534-a161-1dece39e3bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'test_dupe_genes.gtf'\n",
    "df = pr.read_gtf(gtf).df\n",
    "cols = ['Chromosome',\n",
    "             'Feature', \n",
    "             'Start', 'End', \n",
    "             'Score', 'Strand', 'Frame', 'gene_id', 'gene_name', \n",
    "             'gene_status', 'gene_type', 'talon_gene', \n",
    "             'ic', 'ic_id', 'tss_id', 'tss', 'tes_id', 'tes', 'transcript_id',\n",
    "             'transcript_name']\n",
    "gb_cols = list(set(df.columns)&set(cols))\n",
    "real_input = df.copy(deep=True)\n",
    "# print(df.loc[(df.Feature == 'transcript')&(df.transcript_id == 'ENSG00000000460[3,10,3]'), gb_cols])\n",
    "# df = agg_gtf(df)\n",
    "# print()\n",
    "# gb_cols = list(set(df.columns)&set(cols))\n",
    "# print(df.loc[(df.Feature == 'transcript')&(df.transcript_id == 'ENSG00000000460[3,10,3]'), gb_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "25269756-447c-4f74-acf1-56494c4526ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>NNC_transcript</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>ISM_to_IDs</th>\n",
       "      <th>ISM_transcript</th>\n",
       "      <th>ISM-prefix_to_IDs</th>\n",
       "      <th>ISM-prefix_transcript</th>\n",
       "      <th>ISM-suffix_to_IDs</th>\n",
       "      <th>ISM-suffix_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>169794679</td>\n",
       "      <td>169853135</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>169794679</td>\n",
       "      <td>169853135</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>169794679</td>\n",
       "      <td>169853135</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>169794989</td>\n",
       "      <td>169853135</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>169794989</td>\n",
       "      <td>169853135</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2353</th>\n",
       "      <td>chr20</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50934816</td>\n",
       "      <td>50958605</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000079717.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2504</th>\n",
       "      <td>chrX</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>100628619</td>\n",
       "      <td>100636856</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>...</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513</th>\n",
       "      <td>chrX</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>100628619</td>\n",
       "      <td>100636856</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>...</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2564</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>100628619</td>\n",
       "      <td>100636856</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000057483.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>100628619</td>\n",
       "      <td>100636856</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000057483.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>124 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome  Source     Feature      Start        End Score Strand Frame  \\\n",
       "0          chr1   TALON  transcript  169794679  169853135     .      +     .   \n",
       "25         chr1   TALON  transcript  169794679  169853135     .      +     .   \n",
       "50         chr1   TALON  transcript  169794679  169853135     .      +     .   \n",
       "100        chr1   TALON  transcript  169794989  169853135     .      +     .   \n",
       "125        chr1   TALON  transcript  169794989  169853135     .      +     .   \n",
       "...         ...     ...         ...        ...        ...   ...    ...   ...   \n",
       "2353      chr20  HAVANA  transcript   50934816   50958605     .      -     .   \n",
       "2504       chrX   TALON  transcript  100628619  100636856     .      -     .   \n",
       "2513       chrX   TALON  transcript  100628619  100636856     .      -     .   \n",
       "2564       chrX  HAVANA  transcript  100628619  100636856     .      -     .   \n",
       "2573       chrX  HAVANA  transcript  100628619  100636856     .      -     .   \n",
       "\n",
       "                 gene_id gene_name  ... NNC_transcript       transcript_type  \\\n",
       "0     ENSG00000000460.16  C1orf112  ...            NaN                   NaN   \n",
       "25    ENSG00000000460.16  C1orf112  ...            NaN                   NaN   \n",
       "50    ENSG00000000460.16  C1orf112  ...            NaN                   NaN   \n",
       "100   ENSG00000000460.16  C1orf112  ...            NaN                   NaN   \n",
       "125   ENSG00000000460.16  C1orf112  ...            NaN                   NaN   \n",
       "...                  ...       ...  ...            ...                   ...   \n",
       "2353  ENSG00000000419.12      DPM1  ...            NaN  processed_transcript   \n",
       "2504  ENSG00000000003.14    TSPAN6  ...           TRUE                   NaN   \n",
       "2513  ENSG00000000003.14    TSPAN6  ...           TRUE                   NaN   \n",
       "2564  ENSG00000000003.14    TSPAN6  ...            NaN        protein_coding   \n",
       "2573  ENSG00000000003.14    TSPAN6  ...            NaN        protein_coding   \n",
       "\n",
       "     transcript_support_level     havana_transcript ISM_to_IDs ISM_transcript  \\\n",
       "0                         NaN                   NaN        NaN            NaN   \n",
       "25                        NaN                   NaN        NaN            NaN   \n",
       "50                        NaN                   NaN        NaN            NaN   \n",
       "100                       NaN                   NaN        NaN            NaN   \n",
       "125                       NaN                   NaN        NaN            NaN   \n",
       "...                       ...                   ...        ...            ...   \n",
       "2353                        1  OTTHUMT00000079717.1        NaN            NaN   \n",
       "2504                      NaN                   NaN        NaN            NaN   \n",
       "2513                      NaN                   NaN        NaN            NaN   \n",
       "2564                        1  OTTHUMT00000057483.1        NaN            NaN   \n",
       "2573                        1  OTTHUMT00000057483.1        NaN            NaN   \n",
       "\n",
       "     ISM-prefix_to_IDs ISM-prefix_transcript ISM-suffix_to_IDs  \\\n",
       "0                  NaN                   NaN               NaN   \n",
       "25                 NaN                   NaN               NaN   \n",
       "50                 NaN                   NaN               NaN   \n",
       "100                NaN                   NaN               NaN   \n",
       "125                NaN                   NaN               NaN   \n",
       "...                ...                   ...               ...   \n",
       "2353               NaN                   NaN               NaN   \n",
       "2504               NaN                   NaN               NaN   \n",
       "2513               NaN                   NaN               NaN   \n",
       "2564               NaN                   NaN               NaN   \n",
       "2573               NaN                   NaN               NaN   \n",
       "\n",
       "     ISM-suffix_transcript  \n",
       "0                      NaN  \n",
       "25                     NaN  \n",
       "50                     NaN  \n",
       "100                    NaN  \n",
       "125                    NaN  \n",
       "...                    ...  \n",
       "2353                   NaN  \n",
       "2504                   NaN  \n",
       "2513                   NaN  \n",
       "2564                   NaN  \n",
       "2573                   NaN  \n",
       "\n",
       "[124 rows x 46 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[df.Feature=='transcript']\n",
    "temp.loc[temp.transcript_id.duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "482e6862-4f2f-4427-b280-93ff401bd90f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py:5182: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  downcast=downcast,\n",
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n"
     ]
    }
   ],
   "source": [
    "test_agg_gtf(print_dfs=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1ecf767-7333-42d3-a0f6-71129ef07213",
   "metadata": {},
   "source": [
    "## testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "bd7b448d-a5ce-4e07-bc5e-fae62abcc96a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_end_df(c,s,st,e,n, source,mode):\n",
    "    df = pd.DataFrame()\n",
    "    cols = ['Chromosome', 'Strand', 'Start', 'End', 'Name']\n",
    "    var = [c,s,st,e,n]\n",
    "    for col, var in zip(cols, var):\n",
    "        if type(var) == list:\n",
    "            df[col] = var\n",
    "\n",
    "    # add source\n",
    "    df['source'] = source\n",
    "\n",
    "    df = format_end_df(df)\n",
    "\n",
    "    # get end # and gene id\n",
    "    if any(df.Name.isnull()):\n",
    "        df['gene_id'] = np.nan\n",
    "        df[mode] = np.nan\n",
    "    else:\n",
    "        df['gene_id'] = df.Name.str.split('_', expand=True)[0]\n",
    "        df[mode] = df.Name.str.split('_', expand=True)[1]\n",
    "\n",
    "    # get arbitrary unique ids\n",
    "    df['id'] = [i for i in range(len(df.index))]\n",
    "    \n",
    "    # get id\n",
    "    df['{}_id'.format(mode)] = df.gene_id+'_'+df[mode]\n",
    "\n",
    "    return df\n",
    "\n",
    "def format_end_df(df):\n",
    "    sort_cols = ['Chromosome', 'Start', 'End', 'Strand']\n",
    "    df = df.sort_values(by=sort_cols)\n",
    "    order = ['Chromosome', 'Start', 'End', 'Strand', 'Name', 'source']\n",
    "    order = [o for o in order if o in df.columns]\n",
    "    df = df[order]\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    return df\n",
    "\n",
    "def make_exon_df(n,c,e,s,g,t,nt=None,ag1=None,ag2=None):\n",
    "    df = pd.DataFrame()\n",
    "    df['Chromosome'] = c\n",
    "    df['Start'] = [i[0] for i in e]\n",
    "    df['End'] = [i[1] for i in e]\n",
    "    df['Strand'] = s\n",
    "    df['Feature'] = 'exon'\n",
    "    cols = ['gene_name', 'gene_id']\n",
    "    for c in cols:\n",
    "        df[c] = g\n",
    "    cols = ['transcript_id', 'transcript_name']\n",
    "    for c in cols:\n",
    "        df[c] = t\n",
    "    if nt:\n",
    "        df['new_transcript_id'] = nt\n",
    "    if ag1:\n",
    "        df['ag1'] = ag1\n",
    "    if ag2:\n",
    "        df['ag2'] = ag2\n",
    "\n",
    "    # reorder exons and starts/ stops if needed\n",
    "    df['new_Start'] = df[['Start', 'End']].min(axis=1)\n",
    "    df['new_End'] = df[['Start', 'End']].max(axis=1)\n",
    "    df.drop(['Start', 'End'], axis=1, inplace=True)\n",
    "    df.rename({'new_Start':'Start',\n",
    "               'new_End':'End'}, axis=1, inplace=True)\n",
    "    s = s[0]\n",
    "    if s == '+':\n",
    "        ascending = True\n",
    "    elif s == '-':\n",
    "        ascending = False\n",
    "    df.sort_values(by='Start', ascending=ascending, inplace=True)\n",
    "    return df\n",
    "\n",
    "def make_hier_entry(df, how='t'):\n",
    "    \"\"\"\n",
    "    kind {'g','t'}\n",
    "    \"\"\"\n",
    "    agg_dict = {'min_coord': 'min', 'max_coord': 'max'}\n",
    "    t_df = df.copy(deep=True)\n",
    "    t_df['min_coord'] = t_df[['Start', 'End']].min(axis=1)\n",
    "    t_df['max_coord'] = t_df[['Start', 'End']].max(axis=1)\n",
    "    if how == 't':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id', 'transcript_id', 'transcript_name',\n",
    "                   'tss_id', 'tes_id',\n",
    "                   'new_transcript_id', 'original_transcript_id', \n",
    "                   'original_transcript_name', 'ag1', 'ag2']\n",
    "        gb_cols = list(set(gb_cols)&(set(t_df.columns)))\n",
    "    elif how == 'g':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id']\n",
    "        \n",
    "    cols = gb_cols + ['min_coord', 'max_coord']\n",
    "    t_df = t_df[cols]\n",
    "    t_df = t_df.groupby(gb_cols).agg(agg_dict).reset_index()\n",
    "    t_df.rename({'min_coord': 'Start', 'max_coord': 'End'}, axis=1, inplace=True)\n",
    "    if how == 't':\n",
    "        t_df['Feature'] = 'transcript'\n",
    "    elif how == 'g':\n",
    "        t_df['Feature'] = 'gene'\n",
    "        \n",
    "    return t_df\n",
    "\n",
    "def make_test_gtf(ts):\n",
    "    df = pd.concat(ts)\n",
    "        # make transcript entries\n",
    "    t_df = make_hier_entry(df, how='t')\n",
    "    # make gene entries\n",
    "    g_df = make_hier_entry(df, how='g')\n",
    "\n",
    "    # concat everything and sort by gene id, transcript id, feature rank (gene =0, t =1, exon=2), then start coords\n",
    "    df = pd.concat([df, t_df, g_df])\n",
    "    df = sort_gtf(df)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "dd4e126d-fb8b-4ccf-bf85-9e268e0d5d41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tests for agg_gtf\n",
    "def test_agg_gtf(print_dfs=True):\n",
    "    \n",
    "    # need:\n",
    "    # transcripts that don't need to be aggregated\n",
    "    # transcripts that do need to be aggregated\n",
    "\n",
    "    ts = []\n",
    "\n",
    "    # t1 - transcript that won't need to be aggregated\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[1,10], [14,20], [25,30]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    nt = 'g1[1,1,1]'\n",
    "    t = 'g1_t1'\n",
    "    ag1='known'\n",
    "    ag2='p1'\n",
    "    df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t2 - similar transcript from g1 gene that won't need to be aggregated\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[1,10], [14,20], [25,50]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    nt = 'g1[1,1,3]'\n",
    "    t = 'g1_t2'\n",
    "    ag1 = 'novel'\n",
    "    ag2 = 'p1'\n",
    "    df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_3'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t3 - transcript from a different gene that does need to be aggregated\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[90,60], [45,30], [10,8]]\n",
    "    s = ['-' for i in range(n)]\n",
    "    g = 'g2'\n",
    "    nt = 'g2[1,1,1]'\n",
    "    t = 'g2_t1'\n",
    "    ag1 = 'known'\n",
    "    ag2 = 'p3'\n",
    "    df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "    df['tss_id'] = 'g2_1'\n",
    "    df['tes_id'] = 'g2_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t4 - transcript that needs to be collapsed with t3\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[90,60], [45,30], [10,8]]\n",
    "    s = ['-' for i in range(n)]\n",
    "    g = 'g2'\n",
    "    nt = 'g2[1,1,1]'\n",
    "    t = 'g2_t2'\n",
    "    ag1 = 'novel'\n",
    "    ag2 ='p4'\n",
    "    df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "    df['tss_id'] = 'g2_1'\n",
    "    df['tes_id'] = 'g2_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # make test gtf\n",
    "    test_df = make_test_gtf(ts)\n",
    "    test_df.rename({'transcript_id': 'original_transcript_id',\n",
    "                    'transcript_name': 'original_transcript_name'},\n",
    "                   axis=1, inplace=True)\n",
    "    test_df.rename({'new_transcript_id': 'transcript_id'}, \n",
    "                   axis=1, inplace=True)\n",
    "    test_df['transcript_name'] = test_df['transcript_id']\n",
    "\n",
    "    # make ctrl df\n",
    "    ts = []\n",
    "\n",
    "    # t1 - transcript that won't need to be aggregated\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[1,10], [14,20], [25,30]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    t = 'g1[1,1,1]'\n",
    "    ag1='known'\n",
    "    ag2='p1'\n",
    "    df = make_exon_df(n,c,e,s,g,t,ag1=ag1,ag2=ag2)\n",
    "    df['original_transcript_id'] = 'g1_t1'\n",
    "    df['original_transcript_name'] = df['original_transcript_id']\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t2 - similar transcript from g1 gene that won't need to be aggregated\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[1,10], [14,20], [25,50]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    t = 'g1[1,1,3]'\n",
    "    ag1 = 'novel'\n",
    "    ag2 = 'p1'\n",
    "    df = make_exon_df(n,c,e,s,g,t,ag1=ag1,ag2=ag2)\n",
    "    df['original_transcript_id'] = 'g1_t2'\n",
    "    df['original_transcript_name'] = df['original_transcript_id']\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_3'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t3 /t4 collapsed- transcript from a different gene that does need to be aggregated\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[90,60], [45,30], [10,8]]\n",
    "    s = ['-' for i in range(n)]\n",
    "    g = 'g2'\n",
    "    t = 'g2[1,1,1]'\n",
    "    ag1 = 'known,novel'\n",
    "    ag2 = 'p3,p4'\n",
    "    df = make_exon_df(n,c,e,s,g,t,ag1=ag1,ag2=ag2)\n",
    "    df['tss_id'] = 'g2_1'\n",
    "    df['tes_id'] = 'g2_1'\n",
    "    df['original_transcript_id'] = 'g2_t1,g2_t2'\n",
    "    df['original_transcript_name'] = df['original_transcript_id']\n",
    "    ts.append(df)\n",
    "\n",
    "\n",
    "    ctrl = pr.PyRanges(make_test_gtf(ts)).df\n",
    "    test = pr.PyRanges(agg_gtf(test_df)).df\n",
    "\n",
    "    ctrl.reset_index(inplace=True, drop=True)\n",
    "    test.reset_index(inplace=True, drop=True)\n",
    "\n",
    "    if print_dfs:\n",
    "        print('test')\n",
    "        print(test)\n",
    "        print(test.index)\n",
    "        print(test.dtypes)\n",
    "        print('ctrl')\n",
    "        print(ctrl)\n",
    "        print(ctrl.index)\n",
    "        print(ctrl.dtypes)\n",
    "\n",
    "    pd.testing.assert_frame_equal(ctrl, test, check_like=True)\n",
    "\n",
    "    assert len(ctrl.index) == len(test.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "fa9b6b78-9c93-4e35-948a-45398560e282",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tests for update_gtf_ends\n",
    "def test_update_gtf_ends(print_dfs=True):\n",
    "\n",
    "    ts = []\n",
    "    # t1 - fwd strand transcript from gene w/ >2 transcripts\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[1,10], [14,20], [25,30]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    t = 'g1_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_1'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t1.5 - fwd strand transcript from gene that doesn't \n",
    "    # need boundaries updated\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[0, 10], [14,20], [25, 35]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    t = 'g1_t1.5'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t2 - rev. strand transcript from gene w/ >2 transcripts\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[90,60], [45,30], [10,8]]\n",
    "    s = ['-' for i in range(n)]\n",
    "    g = 'g2'\n",
    "    t = 'g2_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g2_1'\n",
    "    df['tes_id'] = 'g2_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t3 - rev. strand transcript from gene w/ >2 transcripts\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[95,60], [45,30], [10,6]]\n",
    "    s = ['-' for i in range(n)]\n",
    "    g = 'g2'\n",
    "    t = 'g2_t2'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g2_2'\n",
    "    df['tes_id'] = 'g2_2'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t4 - fwd strand gene w/ monoexonic transcript\n",
    "    n = 1\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[20,30]]\n",
    "    s = ['+']\n",
    "    g = 'g3'\n",
    "    t = 'g3_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g3_1'\n",
    "    df['tes_id'] = 'g3_1'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t5 - rev strand gene w/ monoexonic transcript\n",
    "    n = 1\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[50, 40]]\n",
    "    s = ['-']\n",
    "    g = 'g4'\n",
    "    t = 'g4_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g4_1'\n",
    "    df['tes_id'] = 'g4_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # tss reference\n",
    "    n = 5\n",
    "    mode = 'tss'\n",
    "    c = ['1' for i in range(n)]\n",
    "    s = ['+', '-', '-', '+', '-']\n",
    "    st = [0, 85, 91, 19, 50] \n",
    "    e = [3, 93, 98, 22, 55]\n",
    "    n = ['g1_1', 'g2_1', 'g2_2', 'g3_1', 'g4_1']\n",
    "    source = 'v1'\n",
    "    tss = make_end_df(c,s,st,e,n,source,mode)\n",
    "    tss = pr.PyRanges(tss)\n",
    "\n",
    "    # tes reference\n",
    "    n = 5\n",
    "    mode = 'tes'\n",
    "    c = ['1' for i in range(n)]\n",
    "    s = ['+', '-', '-', '+', '-']\n",
    "    st = [25, 7, 4, 30, 35]\n",
    "    e = [35, 9, 6, 31, 40]\n",
    "    n = ['g1_1', 'g2_1', 'g2_2', 'g3_1', 'g4_1']\n",
    "    source = 'v1'\n",
    "    tes = make_end_df(c,s,st,e,n,source,mode)\n",
    "    tes = pr.PyRanges(tes)\n",
    "\n",
    "    test_df = make_test_gtf(ts)\n",
    "    # test_df = pr.PyRanges(test_df)\n",
    "\n",
    "    test = update_gtf_ends(test_df, tss, tes)\n",
    "\n",
    "    # ctrl for update_gtf_ends\n",
    "    # tests for update_gtf_ends\n",
    "    ts = []\n",
    "    # t1 - fwd strand transcript\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[0,10], [14,20], [25,35]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    t = 'g1_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_1'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t1.5 - fwd strand transcript from gene that doesn't \n",
    "    # need boundaries updated\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[0, 10], [14,20], [25, 35]]\n",
    "    s = ['+' for i in range(n)]\n",
    "    g = 'g1'\n",
    "    t = 'g1_t1.5'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g1_1'\n",
    "    df['tes_id'] = 'g1_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t2 - rev. strand transcript from gene w/ >2 transcripts\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[93,60], [45,30], [10,7]]\n",
    "    s = ['-' for i in range(n)]\n",
    "    g = 'g2'\n",
    "    t = 'g2_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g2_1'\n",
    "    df['tes_id'] = 'g2_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    # t3 - rev. strand transcript from gene w/ >2 transcripts\n",
    "    n = 3\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[98,60], [45,30], [10,4]]\n",
    "    s = ['-' for i in range(n)]\n",
    "    g = 'g2'\n",
    "    t = 'g2_t2'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g2_2'\n",
    "    df['tes_id'] = 'g2_2'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t4 - fwd strand gene w/ monoexonic transcript\n",
    "    n = 1\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[19,31]]\n",
    "    s = ['+']\n",
    "    g = 'g3'\n",
    "    t = 'g3_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g3_1'\n",
    "    df['tes_id'] = 'g3_1'\n",
    "    ts.append(df)\n",
    "    \n",
    "    # t4 - rev strand gene w/ monoexonic transcript\n",
    "    n = 1\n",
    "    c = ['1' for i in range(n)]\n",
    "    e = [[55, 35]]\n",
    "    s = ['-']\n",
    "    g = 'g4'\n",
    "    t = 'g4_t1'\n",
    "    df = make_exon_df(n,c,e,s,g,t)\n",
    "    df['tss_id'] = 'g4_1'\n",
    "    df['tes_id'] = 'g4_1'\n",
    "    ts.append(df)\n",
    "\n",
    "    ctrl = make_test_gtf(ts)\n",
    "    ctrl.reset_index(inplace=True, drop=True)\n",
    "    test.reset_index(inplace=True, drop=True)\n",
    "\n",
    "    if print_dfs:\n",
    "        print('test')\n",
    "        print(test)\n",
    "        print(test.index)\n",
    "        print(test.dtypes)\n",
    "        print('ctrl')\n",
    "        print(ctrl)\n",
    "        print(ctrl.index)\n",
    "        print(ctrl.dtypes)\n",
    "\n",
    "    pd.testing.assert_frame_equal(ctrl, test, check_like=True)\n",
    "\n",
    "    assert len(ctrl.index) == len(test.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "3bf42937-cc46-4762-9c3e-57500559e2c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_update_gtf_ends(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "8e0ab4fa-925a-46c3-8e1f-13f7dc1d1ac1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tests for map_gtf_ids\n",
    "# - transcript that does not have duplicate\n",
    "# - transcript that does have a duplicate\n",
    "\n",
    "ts = []\n",
    "# t1 - transcript that doesn't need to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[1,10], [14,20], [25,30]]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g1'\n",
    "t = 'g1_t1'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "# t2 - rev. strand transcript that needs to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[90,60], [45,30], [10,8]]\n",
    "s = ['-' for i in range(n)]\n",
    "g = 'g2'\n",
    "t = 'g2_t1'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "# t3 - rev. strand transcript that needs to be merged\n",
    "n = 3\n",
    "c = ['1' for i in range(n)]\n",
    "e = [[95,60], [45,30], [10,6]]\n",
    "s = ['-' for i in range(n)]\n",
    "g = 'g2'\n",
    "t = 'g2_t2'\n",
    "ts.append(make_exon_df(n,c,e,s,g,t))\n",
    "\n",
    "test_df = make_test_gtf(ts)\n",
    "\n",
    "# map file\n",
    "otid = ['g1_t1', 'g2_t1', 'g2_t2',]\n",
    "otname = [o+'n' for o in otid]\n",
    "tid = ['g1[1,1,1]', 'g2[1,1,1]', 'g2[1,1,1]']\n",
    "tname = [o.split('[')[0]+'n['+o.split('[')[1] for o in tid]\n",
    "gid = ['g1', 'g2', 'g2']\n",
    "gname = ['g1n', 'g2n', 'g2n']\n",
    "m_df = pd.DataFrame()\n",
    "m_df['original_transcript_id'] = otid\n",
    "m_df['original_transcript_name'] = otname\n",
    "m_df['transcript_id'] = tid\n",
    "m_df['transcript_name'] = tname\n",
    "m_df['gene_name'] = gname\n",
    "m_df['transcript_triplet'] = m_df.transcript_id.str.slice(2)\n",
    "m_df['gene_id'] = gid\n",
    "\n",
    "# control\n",
    "otid = ['g1_t1', 'g2_t2']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fda3e212-2a3a-45ea-9eb9-877d2a5bee4d",
   "metadata": {},
   "source": [
    "## write h5 ref from input beds and ics tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c97c6fff-29c7-4e14-8d92-37ff668f732e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ic = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/temp/talon_ic.tsv'\n",
    "tes = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tes.bed'\n",
    "tss = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tss.bed'\n",
    "tss_map = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tes_source_map.bed'\n",
    "tes_map = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/lr_bulk/cerberus/test_tss_source_map.bed'\n",
    "\n",
    "# df = read_ic_ref(ic)\n",
    "# df.head()\n",
    "\n",
    "write_reference(tss, tes, ic, 'test.h5')\n",
    "\n",
    "# df = read_cerberus_ends(tss, mode='tss')\n",
    "# df\n",
    "\n",
    "df = read_cerberus_source_map(tss_map)\n",
    "df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
