{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "521791a1-c358-493d-97c3-22ffeb41d263",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ad850dca-ff08-4f1a-89e5-9ebf8734c4bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'tests/files/Canx.gtf'\n",
    "dist = 50\n",
    "\n",
    "gr_gtf = pr.read_gtf(gtf)\n",
    "\n",
    "df_tss = gr_gtf.features.tss()#.extend(dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "de6b1491-9939-4026-93b2-8ab0285ba51f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>tss</td>\n",
       "      <td>50325672</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>tss</td>\n",
       "      <td>50325672</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-001</td>\n",
       "      <td>ENSMUSP00000020637.8</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012410.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>tss</td>\n",
       "      <td>50299377</td>\n",
       "      <td>50299378</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012412.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>tss</td>\n",
       "      <td>50298318</td>\n",
       "      <td>50298319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>tss</td>\n",
       "      <td>50300989</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Source     | Feature    |     Start |       End | Score      | Strand       | +17   |\n",
       "| (category)   | (object)   | (object)   |   (int32) |   (int32) | (object)   | (category)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr11        | ENSEMBL    | tss        |  50325672 |  50325673 | .          | -            | ...   |\n",
       "| chr11        | HAVANA     | tss        |  50325672 |  50325673 | .          | -            | ...   |\n",
       "| chr11        | HAVANA     | tss        |  50299377 |  50299378 | .          | -            | ...   |\n",
       "| chr11        | HAVANA     | tss        |  50298318 |  50298319 | .          | -            | ...   |\n",
       "| chr11        | HAVANA     | tss        |  50300989 |  50300990 | .          | -            | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 5 rows and 24 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Frame, gene_id, gene_type, gene_status, level, havana_gene, transcript_id, transcript_type, ... (+ 9 more.)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tss.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0319f79f-4b3a-4ee1-a2f5-881347e62b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = {'Chromosome': ['chr1', 'chr1', 'chr1'], 'Start': [3, 8, 5], 'End': [6, 9, 7],\n",
    "     'Strand': ['+', '+', '-']}\n",
    "gr = pr.from_dict(d)\n",
    "gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f6e2c6da-c9a9-4043-a998-2feb0f8cdf70",
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_bed = 'tests/files/Canx_tss.bed'\n",
    "tes_bed = 'tests/files/Canx_tes.bed'\n",
    "gtf = 'tests/files/Canx.gtf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "657c0947-23a9-4c29-a114-1404b3a174a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = pr.read_gtf(gtf, duplicate_attr=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "93120be7-6cf8-4f96-9b63-73fddd0ba5f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50325492</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000371309.6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311780</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>50311774</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | Score      | Strand       | +17   |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) | (object)   | (category)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr11        | HAVANA     | gene         |  50293960 |  50325673 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | transcript   |  50293960 |  50325673 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50325492 |  50325673 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50311603 |  50311780 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | CDS          |  50311603 |  50311777 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | start_codon  |  50311774 |  50311777 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50310793 |  50310867 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | CDS          |  50310793 |  50310867 | .          | -            | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 8 rows and 24 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Frame, gene_id, gene_type, gene_status, level, havana_gene, transcript_id, transcript_type, ... (+ 9 more.)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "90e2a7c7-2006-4a83-87a4-0335052cf8b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = gtf.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "be7653a4-03cd-4f71-8ad7-bde3d009f797",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = gtf.loc[gtf.Feature == 'transcript']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a6ed3fda-e718-4ce3-8a7f-2bed56cd475f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50294466</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-001</td>\n",
       "      <td>ENSMUSP00000020637.8</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012410.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50296198</td>\n",
       "      <td>50299378</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012412.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50297363</td>\n",
       "      <td>50298319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome   Source     Feature     Start       End Score Strand Frame  \\\n",
       "1       chr11  ENSEMBL  transcript  50293960  50325673     .      -     .   \n",
       "36      chr11   HAVANA  transcript  50294466  50325673     .      -     .   \n",
       "71      chr11   HAVANA  transcript  50296198  50299378     .      -     .   \n",
       "77      chr11   HAVANA  transcript  50297363  50298319     .      -     .   \n",
       "80      chr11   HAVANA  transcript  50298837  50300990     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ...       transcript_type  \\\n",
       "1   ENSMUSG00000020368.14  protein_coding  ...        protein_coding   \n",
       "36  ENSMUSG00000020368.14  protein_coding  ...        protein_coding   \n",
       "71  ENSMUSG00000020368.14  protein_coding  ...  processed_transcript   \n",
       "77  ENSMUSG00000020368.14  protein_coding  ...  processed_transcript   \n",
       "80  ENSMUSG00000020368.14  protein_coding  ...  processed_transcript   \n",
       "\n",
       "   transcript_status transcript_name            protein_id  \\\n",
       "1              KNOWN        Canx-201  ENSMUSP00000137440.1   \n",
       "36             KNOWN        Canx-001  ENSMUSP00000020637.8   \n",
       "71             KNOWN        Canx-003                   NaN   \n",
       "77             KNOWN        Canx-002                   NaN   \n",
       "80             KNOWN        Canx-004                   NaN   \n",
       "\n",
       "                              tag transcript_support_level       ccdsid  \\\n",
       "1   basic,appris_principal_1,CCDS                        1  CCDS24633.1   \n",
       "36  basic,appris_principal_1,CCDS                        1  CCDS24633.1   \n",
       "71                            NaN                        2          NaN   \n",
       "77                            NaN                        5          NaN   \n",
       "80                            NaN                        3          NaN   \n",
       "\n",
       "   exon_number exon_id     havana_transcript  \n",
       "1          NaN     NaN                   NaN  \n",
       "36         NaN     NaN  OTTMUST00000012410.1  \n",
       "71         NaN     NaN  OTTMUST00000012412.1  \n",
       "77         NaN     NaN  OTTMUST00000012411.1  \n",
       "80         NaN     NaN  OTTMUST00000012413.1  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "45b628dc-6f5a-4b0f-8b7d-e053e0f15b53",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf['basic_set'] = gtf.tag.str.contains('basic')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "12b1093c-f446-4408-ab45-c901de7bf53f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf['temp'] = gtf.tag.str.split('appris_principal_', n=1, expand=True)[1]\n",
    "gtf['appris_principal'] = gtf.temp.str.split(',', n=1, expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "972aa359-e0b4-4725-8389-b1cadbe4fdcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf.basic_set.fillna(False, inplace=True)\n",
    "gtf = gtf.sort_values(by=['appris_principal', 'basic_set'], \n",
    "                      ascending=[False, True],\n",
    "                      na_position='last')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "0abfe0c1-3e20-4c79-97eb-f84bd97dbf5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>basic_set</th>\n",
       "      <th>temp</th>\n",
       "      <th>appris_principal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>1,CCDS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50294466</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSP00000020637.8</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012410.1</td>\n",
       "      <td>True</td>\n",
       "      <td>1,CCDS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50296198</td>\n",
       "      <td>50299378</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012412.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50297363</td>\n",
       "      <td>50298319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome   Source     Feature     Start       End Score Strand Frame  \\\n",
       "1       chr11  ENSEMBL  transcript  50293960  50325673     .      -     .   \n",
       "36      chr11   HAVANA  transcript  50294466  50325673     .      -     .   \n",
       "71      chr11   HAVANA  transcript  50296198  50299378     .      -     .   \n",
       "77      chr11   HAVANA  transcript  50297363  50298319     .      -     .   \n",
       "80      chr11   HAVANA  transcript  50298837  50300990     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ...            protein_id  \\\n",
       "1   ENSMUSG00000020368.14  protein_coding  ...  ENSMUSP00000137440.1   \n",
       "36  ENSMUSG00000020368.14  protein_coding  ...  ENSMUSP00000020637.8   \n",
       "71  ENSMUSG00000020368.14  protein_coding  ...                   NaN   \n",
       "77  ENSMUSG00000020368.14  protein_coding  ...                   NaN   \n",
       "80  ENSMUSG00000020368.14  protein_coding  ...                   NaN   \n",
       "\n",
       "                              tag transcript_support_level       ccdsid  \\\n",
       "1   basic,appris_principal_1,CCDS                        1  CCDS24633.1   \n",
       "36  basic,appris_principal_1,CCDS                        1  CCDS24633.1   \n",
       "71                            NaN                        2          NaN   \n",
       "77                            NaN                        5          NaN   \n",
       "80                            NaN                        3          NaN   \n",
       "\n",
       "   exon_number exon_id     havana_transcript basic_set    temp  \\\n",
       "1          NaN     NaN                   NaN      True  1,CCDS   \n",
       "36         NaN     NaN  OTTMUST00000012410.1      True  1,CCDS   \n",
       "71         NaN     NaN  OTTMUST00000012412.1     False     NaN   \n",
       "77         NaN     NaN  OTTMUST00000012411.1     False     NaN   \n",
       "80         NaN     NaN  OTTMUST00000012413.1     False     NaN   \n",
       "\n",
       "   appris_principal  \n",
       "1                 1  \n",
       "36                1  \n",
       "71              NaN  \n",
       "77              NaN  \n",
       "80              NaN  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb0f1756-e68e-434f-8560-086a7945b188",
   "metadata": {},
   "source": [
    "## compute triplets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "003d5746-81e3-4c99-b32c-ef798294b716",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transcript_ref(fname):\n",
    "    \n",
    "    df = pr.read_gtf(fname, duplicate_attr=True).df\n",
    "    df = df.loc[df.Feature == 'transcript']\n",
    "    df['basic_set'] = df.tag.str.contains('basic')\n",
    "    df['temp'] = df.tag.str.split('appris_principal_', n=1, expand=True)[1]\n",
    "    df['appris_principal'] = df.temp.str.split(',', n=1, expand=True)[0]\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "id": "dbf0861b-8d52-4452-bafa-75a5434f61bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def number_tss_ic_tes(df, mode):\n",
    "    # groupby feature but record which feature\n",
    "    # each transcript id uses \n",
    "    cols = ['transcript_id', mode, 'basic_set', 'appris_principal', 'gene_id']\n",
    "    df = df[cols].groupby([mode, 'gene_id']).agg({'transcript_id': ','.join,\n",
    "                                     'basic_set': 'max',\n",
    "                                     'appris_principal': 'min'}).reset_index()\n",
    "\n",
    "    # compute intron chain number\n",
    "    df['{}_num'.format(mode)] = df.sort_values(by=['gene_id', 'appris_principal', 'basic_set'],\n",
    "                                 ascending=[True, False, True],\n",
    "                                 na_position='last')\\\n",
    "                                 .groupby(['gene_id'])\\\n",
    "                                 .cumcount() + 1\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "id": "b8c2ab21-29ee-498d-aeef-604be3aebd92",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'tests/files/Canx.gtf'\n",
    "tss_bed = 'tests/files/Canx_tss.bed'\n",
    "tes_bed = 'tests/files/Canx_tes.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "id": "33020490-415a-4ae3-ab13-43ac345164ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "### intron chain annotation ###\n",
    "\n",
    "# get basic status and appris_principal tag for each transcript\n",
    "t_df = get_transcript_ref(fname)\n",
    "t_df = t_df[['transcript_id', 'gene_id', 'basic_set', 'appris_principal']]\n",
    "\n",
    "# get unique intron chains from gtf\n",
    "df = pr.read_gtf(fname).df\n",
    "\n",
    "# restrict to exon entries\n",
    "df = df.loc[df.Feature == 'exon']\n",
    "cols = ['Chromosome', 'Strand', 'Start', 'End', 'transcript_id', 'gene_id']\n",
    "df = df[cols]\n",
    "\n",
    "# melt to isolate individual coordinates\n",
    "df = pd.melt(df, id_vars=['Chromosome', 'Strand', 'transcript_id', 'gene_id'],\n",
    "            value_vars=['Start', 'End'],\n",
    "            value_name='Coord')\n",
    "df.drop('variable', axis=1, inplace=True)\n",
    "\n",
    "# sort to order coordinates correctly\n",
    "df.Coord = df.Coord.astype(int)\n",
    "fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "\n",
    "fwd.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                ascending=[True, True, True])\n",
    "rev.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                ascending=[True, True, False])\n",
    "df = pd.concat([fwd, rev])\n",
    "\n",
    "# create intron chain strings\n",
    "df.Coord = df.Coord.astype(str)\n",
    "df = df.groupby(['Chromosome', 'Strand', 'transcript_id'])['Coord'].apply('-'.join).reset_index()\n",
    "\n",
    "# remove tss and tes from intron chain\n",
    "df['temp'] = df.Coord.str.split('-', n=1, expand=True)[1]\n",
    "df['ic'] = df.temp.str.rsplit('-', n=1, expand=True)[0]\n",
    "\n",
    "# add basic annotation, appris principal number, and gene id\n",
    "df = df.merge(t_df, on='transcript_id', how='left')\n",
    "\n",
    "# add number for each unique intron chain\n",
    "df = number_tss_ic_tes(df, mode='ic')\n",
    "\n",
    "# keep track of which transcripts use each intron chain\n",
    "df.transcript_id = df.transcript_id.str.split(',')\n",
    "ttrip_df = df[['transcript_id', 'ic_num']].explode(column='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "id": "b82ba4e2-8ed5-41bb-b99f-b9028177d7c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### tss / tes numbering ###\n",
    "beds = {}\n",
    "for mode, bed_file in zip(['tss', 'tes'], [tss_bed, tes_bed]):\n",
    "\n",
    "    ends = pr.read_bed(bed_file).df\n",
    "    ends.rename({'ThickStart': 'gene_id'}, axis=1, inplace=True)\n",
    "    ends[mode] = [i for i in range(len(ends.index))]\n",
    "    ends = pr.PyRanges(ends)\n",
    "\n",
    "    # merge gtf ends with bed ends\n",
    "    df = pr.read_gtf(gtf)\n",
    "    if mode == 'tss':\n",
    "        df = df.features.tss().df\n",
    "    elif mode == 'tes':\n",
    "        df = df.features.tes().df\n",
    "\n",
    "    # merge in information about transcript status\n",
    "    t_df = get_transcript_ref(gtf)\n",
    "    t_df = t_df[['transcript_id', 'appris_principal', 'basic_set']]\n",
    "    df = df.merge(t_df, how='left', on='transcript_id')\n",
    "    df = pr.PyRanges(df)\n",
    "\n",
    "    # join with bed ends\n",
    "    df = df.join(ends, \n",
    "                 strandedness='same',\n",
    "                 how='left').df\n",
    "    df = df.loc[df.gene_id == df.gene_id_b]\n",
    "\n",
    "    # add number for each unique end region\n",
    "    df = number_tss_ic_tes(df, mode=mode)\n",
    "    \n",
    "    # keep track of which transcripts use each intron chain\n",
    "    df.transcript_id = df.transcript_id.str.split(',')\n",
    "    temp = df[['transcript_id', '{}_num'.format(mode)]].explode(column='transcript_id')  \n",
    "    ttrip_df = ttrip_df.merge(temp, on='transcript_id')   \n",
    "    \n",
    "    # add numbers back into bed file\n",
    "    ends = ends.df\n",
    "    ends = ends.merge(df[[mode, '{}_num'.format(mode)]], on=mode)\n",
    "    \n",
    "    ends['Name'] = ends.gene_id+'_'+ends['{}_num'.format(mode)].astype(str)\n",
    "    ends.drop([mode, 'gene_id', '{}_num'.format(mode)], axis=1, inplace=True)\n",
    "    beds[mode] = pr.PyRanges(ends)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "id": "d4e810df-c35b-43e9-858d-32dd3951292c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create formatted string to represent each transcript\n",
    "ttrip_df['transcript_triplet'] = '['+ttrip_df['tss_num'].astype(str)+', '+\\\n",
    "                                     ttrip_df['ic_num'].astype(str)+', '+\\\n",
    "                                     ttrip_df['tes_num'].astype(str)+']'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "59900b0b-a160-4684-b57f-ab69922009bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ic_num</th>\n",
       "      <th>tss_num</th>\n",
       "      <th>tes_num</th>\n",
       "      <th>transcript_triplet</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>[4, 4, 5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>[3, 5, 4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>[2, 1, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000155801.2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>[2, 1, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>[1, 3, 2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  ic_num  tss_num  tes_num transcript_triplet\n",
       "0  ENSMUST00000153068.1       4        4        5          [4, 4, 5]\n",
       "1  ENSMUST00000146979.1       5        3        4          [3, 5, 4]\n",
       "2  ENSMUST00000155801.1       1        2        3          [2, 1, 3]\n",
       "3  ENSMUST00000155801.2       1        2        3          [2, 1, 3]\n",
       "4  ENSMUST00000179865.6       2        1        1          [1, 2, 1]\n",
       "5  ENSMUST00000020637.8       3        1        2          [1, 3, 2]"
      ]
     },
     "execution_count": 354,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ttrip_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "id": "00935205-a74a-4f3c-aa64-6eaa3755a825",
   "metadata": {},
   "outputs": [],
   "source": [
    "### output files ###\n",
    "\n",
    "# gtf\n",
    "df = pr.read_gtf(gtf, duplicate_attr=True).df\n",
    "df = df.merge(ttrip_df[['transcript_id', 'transcript_triplet']], \n",
    "              how='left',\n",
    "              on='transcript_id')\n",
    "df = pr.PyRanges(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "id": "a371d9ba-f01b-485b-ae1a-e80cf99c62dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50325622</td>\n",
       "      <td>50325723</td>\n",
       "      <td>ENSMUSG00000020368.14_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50299327</td>\n",
       "      <td>50299428</td>\n",
       "      <td>ENSMUSG00000020368.14_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50298268</td>\n",
       "      <td>50298369</td>\n",
       "      <td>ENSMUSG00000020368.14_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50300939</td>\n",
       "      <td>50301040</td>\n",
       "      <td>ENSMUSG00000020368.14_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------------------+------------+--------------+\n",
       "| Chromosome   |     Start |       End | Name                    | Score      | Strand       |\n",
       "| (category)   |   (int32) |   (int32) | (object)                | (object)   | (category)   |\n",
       "|--------------+-----------+-----------+-------------------------+------------+--------------|\n",
       "| chr11        |  50325622 |  50325723 | ENSMUSG00000020368.14_1 | .          | -            |\n",
       "| chr11        |  50299327 |  50299428 | ENSMUSG00000020368.14_3 | .          | -            |\n",
       "| chr11        |  50298268 |  50298369 | ENSMUSG00000020368.14_4 | .          | -            |\n",
       "| chr11        |  50300939 |  50301040 | ENSMUSG00000020368.14_2 | .          | -            |\n",
       "+--------------+-----------+-----------+-------------------------+------------+--------------+\n",
       "Stranded PyRanges object has 4 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 356,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beds['tss']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "id": "2ddeb85b-5531-4f3a-a720-f779749b640d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>transcript_triplet</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50325492</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000371309.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311780</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>50311774</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | Score      | +19   |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) | (object)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+-------|\n",
       "| chr11        | HAVANA     | gene         |  50293960 |  50325673 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | transcript   |  50293960 |  50325673 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50325492 |  50325673 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50311603 |  50311780 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | CDS          |  50311603 |  50311777 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | start_codon  |  50311774 |  50311777 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50310793 |  50310867 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | CDS          |  50310793 |  50310867 | .          | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+-------+\n",
       "Stranded PyRanges object has 8 rows and 25 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "19 hidden columns: Strand, Frame, gene_id, gene_type, gene_status, level, havana_gene, transcript_id, ... (+ 11 more.)"
      ]
     },
     "execution_count": 338,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "id": "f3309e1d-2e1e-44b1-90d1-c391a4a06ac2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "94c00c48-876a-4803-bd47-65f5846a8ba7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ic</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>basic_set</th>\n",
       "      <th>appris_principal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>50297363-50298319</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>50298116-50297273-50297057-50296198-50299378-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>50298837-50300990</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>50311603-50310793-50310371-50308787-50308309-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>50311603-50310793-50310371-50308787-50308309-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  ic                gene_id  \\\n",
       "0                                  50297363-50298319  ENSMUSG00000020368.14   \n",
       "1  50298116-50297273-50297057-50296198-50299378-5...  ENSMUSG00000020368.14   \n",
       "2                                  50298837-50300990  ENSMUSG00000020368.14   \n",
       "3  50311603-50310793-50310371-50308787-50308309-5...  ENSMUSG00000020368.14   \n",
       "4  50311603-50310793-50310371-50308787-50308309-5...  ENSMUSG00000020368.14   \n",
       "\n",
       "                               transcript_id basic_set appris_principal  \n",
       "0                       ENSMUST00000153068.1       NaN              NaN  \n",
       "1                       ENSMUST00000146979.1       NaN              NaN  \n",
       "2  ENSMUST00000155801.1,ENSMUST00000155801.2      True                1  \n",
       "3                       ENSMUST00000179865.6      True                1  \n",
       "4                       ENSMUST00000020637.8      True                1  "
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "c64e4149-01f0-46fc-992f-a774b84a7945",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "eb44df97-2164-4160-9960-39670f6bd887",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ic</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>basic_set</th>\n",
       "      <th>appris_principal</th>\n",
       "      <th>ic_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>50297363-50298319</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>50298116-50297273-50297057-50296198-50299378-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>50298837-50300990</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>50311603-50310793-50310371-50308787-50308309-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>50311603-50310793-50310371-50308787-50308309-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  ic                gene_id  \\\n",
       "0                                  50297363-50298319  ENSMUSG00000020368.14   \n",
       "1  50298116-50297273-50297057-50296198-50299378-5...  ENSMUSG00000020368.14   \n",
       "2                                  50298837-50300990  ENSMUSG00000020368.14   \n",
       "3  50311603-50310793-50310371-50308787-50308309-5...  ENSMUSG00000020368.14   \n",
       "4  50311603-50310793-50310371-50308787-50308309-5...  ENSMUSG00000020368.14   \n",
       "\n",
       "                               transcript_id basic_set appris_principal  \\\n",
       "0                       ENSMUST00000153068.1       NaN              NaN   \n",
       "1                       ENSMUST00000146979.1       NaN              NaN   \n",
       "2  ENSMUST00000155801.1,ENSMUST00000155801.2      True                1   \n",
       "3                       ENSMUST00000179865.6      True                1   \n",
       "4                       ENSMUST00000020637.8      True                1   \n",
       "\n",
       "   ic_num  \n",
       "0       4  \n",
       "1       5  \n",
       "2       1  \n",
       "3       2  \n",
       "4       3  "
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "018e6889-01e6-4db1-bdca-ea337fe2aefa",
   "metadata": {},
   "source": [
    "#### make canx abundance file for testing purposes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "id": "6c1b35a5-a01c-4260-8c13-7b1fe6cfe339",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/Users/fairliereese/mortazavi_lab/data/mousewg/lr_bulk/talon/mouse_talon_abundance.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 418,
   "id": "72676460-b8df-43b6-b87a-c4a827d4bf1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.loc[df.annot_gene_name == 'Canx']\n",
    "df = df.loc[df.transcript_novelty == 'Known']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 419,
   "id": "10070474-4b6a-49e3-a9d4-d87b4b19d65c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename transcripts according to transcript names\n",
    "gtf_df = pr.read_gtf('tests/files/Canx.gtf').df\n",
    "gtf_df = gtf_df.loc[gtf_df.Feature == 'transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 420,
   "id": "26a64c27-e35c-4ef2-a67d-a191e6ac2e10",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df.transcript_id\n",
    "gtf_df['tid_stable'] = gtf_df.transcript_id.str.split('.', n=1, expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 421,
   "id": "d3ebd5f7-a27c-4d4a-9ee5-331280014393",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['tid_stable'] = df.annot_transcript_id.str.split('.', n=1, expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 422,
   "id": "6e4e8ac8-e431-40d0-b8a6-c5315eebe979",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = gtf_df[['tid_stable', 'transcript_id']]\n",
    "df = df.merge(gtf_df, on='tid_stable')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 423,
   "id": "cf8bcbb7-39db-4880-ab38-3b425efdb2f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.annot_transcript_id = df.transcript_id\n",
    "df.drop(['transcript_id', 'tid_stable'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 424,
   "id": "105abb94-3272-44b0-8f5b-fd2bfe3f6dbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>gastroc_25d_m_2</th>\n",
       "      <th>gastroc_25d_m_1</th>\n",
       "      <th>gastroc_10d_f_2</th>\n",
       "      <th>gastroc_10d_f_1</th>\n",
       "      <th>gastroc_25d_f_1</th>\n",
       "      <th>gastroc_25d_f_2</th>\n",
       "      <th>gastroc_4d_m_1</th>\n",
       "      <th>gastroc_4d_m_2</th>\n",
       "      <th>gastroc_10d_m_1</th>\n",
       "      <th>gastroc_10d_m_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>32839</td>\n",
       "      <td>86107</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-205</td>\n",
       "      <td>15</td>\n",
       "      <td>4281</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32839</td>\n",
       "      <td>86108</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>15</td>\n",
       "      <td>3779</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>44</td>\n",
       "      <td>26</td>\n",
       "      <td>67</td>\n",
       "      <td>38</td>\n",
       "      <td>35</td>\n",
       "      <td>104</td>\n",
       "      <td>96</td>\n",
       "      <td>60</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>32839</td>\n",
       "      <td>86109</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-202</td>\n",
       "      <td>5</td>\n",
       "      <td>708</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32839</td>\n",
       "      <td>86110</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-203</td>\n",
       "      <td>2</td>\n",
       "      <td>234</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32839</td>\n",
       "      <td>86111</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-204</td>\n",
       "      <td>2</td>\n",
       "      <td>600</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>32839</td>\n",
       "      <td>86111</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000155801.2</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-204</td>\n",
       "      <td>2</td>\n",
       "      <td>600</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 102 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_ID  transcript_ID          annot_gene_id   annot_transcript_id  \\\n",
       "0    32839          86107  ENSMUSG00000020368.15  ENSMUST00000179865.6   \n",
       "1    32839          86108  ENSMUSG00000020368.15  ENSMUST00000020637.8   \n",
       "2    32839          86109  ENSMUSG00000020368.15  ENSMUST00000146979.1   \n",
       "3    32839          86110  ENSMUSG00000020368.15  ENSMUST00000153068.1   \n",
       "4    32839          86111  ENSMUSG00000020368.15  ENSMUST00000155801.1   \n",
       "5    32839          86111  ENSMUSG00000020368.15  ENSMUST00000155801.2   \n",
       "\n",
       "  annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "0            Canx              Canx-205       15    4281        Known   \n",
       "1            Canx              Canx-201       15    3779        Known   \n",
       "2            Canx              Canx-202        5     708        Known   \n",
       "3            Canx              Canx-203        2     234        Known   \n",
       "4            Canx              Canx-204        2     600        Known   \n",
       "5            Canx              Canx-204        2     600        Known   \n",
       "\n",
       "  transcript_novelty  ... gastroc_25d_m_2  gastroc_25d_m_1  gastroc_10d_f_2  \\\n",
       "0              Known  ...               1                0                0   \n",
       "1              Known  ...              12               44               26   \n",
       "2              Known  ...               1                1                1   \n",
       "3              Known  ...               0                0                0   \n",
       "4              Known  ...               0                0                0   \n",
       "5              Known  ...               0                0                0   \n",
       "\n",
       "   gastroc_10d_f_1  gastroc_25d_f_1  gastroc_25d_f_2  gastroc_4d_m_1  \\\n",
       "0                3                0                1               4   \n",
       "1               67               38               35             104   \n",
       "2                1                0                2               2   \n",
       "3                0                0                0               0   \n",
       "4                0                0                0               0   \n",
       "5                0                0                0               0   \n",
       "\n",
       "   gastroc_4d_m_2  gastroc_10d_m_1  gastroc_10d_m_2  \n",
       "0               2                0                0  \n",
       "1              96               60               34  \n",
       "2               1                3                0  \n",
       "3               0                0                0  \n",
       "4               0                0                0  \n",
       "5               0                0                0  \n",
       "\n",
       "[6 rows x 102 columns]"
      ]
     },
     "execution_count": 424,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "id": "4294d336-7070-4af9-88e3-006accfbe8b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('tests/files/Canx_abundance.tsv', sep='\\t', index=False)\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eee194e-e209-4da4-8cdc-44f2f69ba314",
   "metadata": {},
   "source": [
    "#### Changing transcript IDs in TALON abundance file and optionally? aggregating counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4e6045ba-5775-4e5c-b95c-547c324fc149",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nov_ranks():\n",
    "    \"\"\"\n",
    "    Get rank for each novelty category\n",
    "    \n",
    "    Returns:\n",
    "        nov_rank (dict): Dictionary of novelty category: int rank\n",
    "        rank_nov (dict): Dictionary of int rank: novelty category\n",
    "    \"\"\"\n",
    "    nov_rank = {'Known': 0,\n",
    "           'NIC': 1,\n",
    "           'ISM': 2,\n",
    "           'NNC': 3,\n",
    "           'Genomic': 4,\n",
    "           'Antisense': 5,\n",
    "           'Intergenic': 6}\n",
    "    \n",
    "    rank_nov = dict()\n",
    "    for key, item in nov_rank.items():\n",
    "        rank_nov[item] = key\n",
    "\n",
    "    return nov_rank, rank_nov\n",
    "\n",
    "def get_ism_ranks():\n",
    "    \"\"\"\n",
    "    Get rank for each ISM subtype category \n",
    "    \n",
    "    Returns:\n",
    "        ism_rank (dict): Dictionary of ism subtype: int rank\n",
    "        rank_ism (dict): Dictionary of int rank: ism subtype\n",
    "    \"\"\"\n",
    "    ism_rank = {'Both': 0,\n",
    "                'Prefix': 1,\n",
    "                'Suffix': 2,\n",
    "                None: 3}\n",
    "    rank_ism = dict()\n",
    "    for key, item in ism_rank.items():\n",
    "        rank_ism[item] = key\n",
    "\n",
    "    return ism_rank, rank_ism\n",
    "\n",
    "def get_non_dataset_cols():\n",
    "    \"\"\"\n",
    "    Get a list of ordered columns from TALON abundance file that\n",
    "    are the non-dataset columns\n",
    "    \n",
    "    Returns:\n",
    "        non_dataset_cols (list of str): List of non-dataset columns\n",
    "    \"\"\"\n",
    "    non_dataset_cols = ['gene_ID', 'transcript_ID', 'annot_gene_id',\n",
    "                   'annot_transcript_id', 'annot_gene_name',\n",
    "                   'annot_transcript_name', 'n_exons', 'length',\n",
    "                   'gene_novelty', 'transcript_novelty', 'ISM_subtype']\n",
    "    return non_dataset_cols\n",
    "\n",
    "def get_dataset_cols(df):\n",
    "    \"\"\"\n",
    "    Get a list of the dataset columns from a TALON abundance file\n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): TALON abundance file\n",
    "        \n",
    "    Returns:\n",
    "        dataset_cols (list of str): List of dataset columns\n",
    "    \"\"\"\n",
    "    non_dataset_cols = get_non_dataset_cols()\n",
    "    dataset_cols = [ x for x in list(df.columns) \\\n",
    "                        if x not in non_dataset_cols ]\n",
    "    return dataset_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "a65c1ff5-f348-4092-bc2f-d498b6f7a6c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def replace_ab_ids(ab, m, agg):\n",
    "    \"\"\"\n",
    "    Replace the transcript ids and transcript names in a TALON abundance file\n",
    "    with the new transcript ids that contain the triplet\n",
    "    \n",
    "    Parameters:\n",
    "        ab (str): Path to TALON abundance file\n",
    "        m (str): Path to map file (output from assign_triplets)\n",
    "        agg (bool): Aggregate / collapse transcripts with the same triplets\n",
    "            and sum up their count values\n",
    "        \n",
    "    Returns:\n",
    "        df (pandas DataFrame): TALON abundance file with updated\n",
    "            transcript ids / names\n",
    "    \"\"\"\n",
    "    df = pd.read_csv(ab, sep='\\t')\n",
    "    m_df = pd.read_csv(m, sep='\\t')\n",
    "\n",
    "    # fix transcript ids in abundance file\n",
    "    ab_map = m_df[['original_transcript_id', 'original_transcript_name', 'transcript_name', 'transcript_id']]\n",
    "    df = df.merge(ab_map, left_on='annot_transcript_id', right_on='original_transcript_id')\n",
    "\n",
    "    df.drop(['original_transcript_id', 'original_transcript_name', \n",
    "             'annot_transcript_id', 'annot_transcript_name'],\n",
    "            axis=1, inplace=True)\n",
    "    df.rename({'transcript_name': 'annot_transcript_name',\n",
    "               'transcript_id': 'annot_transcript_id'},\n",
    "              axis=1, inplace=True)\n",
    "\n",
    "    # aggregate counts if requested\n",
    "    if agg:\n",
    "        gb_cols = ['gene_ID', 'annot_gene_id', 'annot_gene_name',\n",
    "                   'gene_novelty', 'annot_transcript_name',\n",
    "                   'annot_transcript_id']\n",
    "\n",
    "        # handle properties which won't always correspond across the transcripts\n",
    "        # these are all subject to change\n",
    "        nov_rank, rank_nov = get_nov_ranks()\n",
    "        df['nov_rank'] = df.transcript_novelty.map(nov_rank)\n",
    "        df.drop('transcript_novelty', axis=1, inplace=True)\n",
    "\n",
    "        ism_rank, rank_ism = get_ism_ranks()\n",
    "        df['ism_rank'] = df.ISM_subtype.map(ism_rank)\n",
    "        df.drop('ISM_subtype', axis=1, inplace=True)\n",
    "\n",
    "        df['transcript_ID'] = df.transcript_ID.astype(str)\n",
    "        agg_dict = {'transcript_ID': ','.join,\n",
    "                    'n_exons': 'mean', \n",
    "                    'length': 'mean',\n",
    "                    'nov_rank': 'min',\n",
    "                    'ism_rank': 'min'}\n",
    "        cols = gb_cols + list(agg_dict.keys())\n",
    "        for c in list(set(df.columns)-set(cols)):\n",
    "            agg_dict[c] = 'sum'\n",
    "\n",
    "        df = df.groupby(gb_cols).agg(agg_dict).reset_index()\n",
    "        df['transcript_novelty'] = df.nov_rank.map(rank_nov)\n",
    "        df.drop(['nov_rank'], axis=1, inplace=True)\n",
    "        df['ISM_subtype'] = df.ism_rank.map(rank_ism)\n",
    "        df.drop(['ism_rank'], axis=1, inplace=True)\n",
    "\n",
    "    # reorder columns\n",
    "    c1 = get_non_dataset_cols()\n",
    "    c2 = get_dataset_cols(df)\n",
    "    df = df[c1+c2]\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "af9f5ad4-441c-4289-b9c8-1cbc7ef144d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def replace_gtf_ids(gtf, m, agg):\n",
    "    \"\"\"\n",
    "    Replace transcript ids and names in a gtf with the triplets\n",
    "    calculated from assign_triplets\n",
    "    \n",
    "    Parameters:\n",
    "        gtf (str): Path to gtf file\n",
    "        m (str): Path to map file output from assign_triplets\n",
    "        agg (bool): Whether or not to collapse transcripts with\n",
    "            duplicate triplets\n",
    "            \n",
    "    Returns:\n",
    "        df (pyranges PyRanges): PyRanges gtf table with updated ids\n",
    "    \"\"\"\n",
    "    \n",
    "    df = pr.read_gtf(gtf).df\n",
    "    m_df = pd.read_csv(m, sep='\\t')\n",
    "\n",
    "    # groupby transcripts that are the same\n",
    "    gb_cols = ['gene_name', 'gene_id', 'transcript_triplet',\n",
    "               'transcript_id', 'transcript_name']\n",
    "    temp = m_df[['transcript_id',\n",
    "                 'original_transcript_id',\n",
    "                 'original_transcript_name']].copy(deep=True)\n",
    "    m_df = m_df.groupby(gb_cols).agg({'original_transcript_id': ','.join,\n",
    "                                      'original_transcript_name': ','.join}).reset_index()\n",
    "    m_df = m_df.merge(temp, on='transcript_id', suffixes=('','_merge'))\n",
    "    m_df.drop(['gene_name', 'gene_id', 'transcript_triplet'],\n",
    "              axis=1, inplace=True)\n",
    "\n",
    "    # add new transcript ids \n",
    "    df = df.merge(m_df, left_on=['transcript_id', 'transcript_name'],\n",
    "                  right_on=['original_transcript_id_merge',\n",
    "                            'original_transcript_name_merge'],\n",
    "                 suffixes=('_x', ''))\n",
    "\n",
    "    # drop old tids\n",
    "    df.drop(['transcript_id_x', 'transcript_name_x',\n",
    "             'original_transcript_name_merge'],\n",
    "            axis=1, inplace=True)\n",
    "\n",
    "    # remove duplicated transcripts; just keeping the first one\n",
    "    if agg:\n",
    "        temp = df[['transcript_id', 'original_transcript_id_merge']].drop_duplicates()\n",
    "        dupe_old_tids = temp.loc[temp.transcript_id.duplicated(keep='first'), 'original_transcript_id_merge']\n",
    "        df = df.loc[~df.original_transcript_id_merge.isin(dupe_old_tids)]  \n",
    "\n",
    "    # drop last column\n",
    "    df.drop('original_transcript_id_merge', axis=1, inplace=True)\n",
    "\n",
    "    df = pr.PyRanges(df)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "16cb0ae6-ac7f-4bb8-9461-ecf54bdceed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = 'tests/files/Canx_abundance.tsv'\n",
    "gtf = 'tests/files/Canx.gtf'\n",
    "m = 'tests/files/Canx_triplet_tid_map.tsv'\n",
    "h5 = '' # option to use / save in previous step as h5df?\n",
    "\n",
    "beep =  replace_ab_ids(ab, m, True)\n",
    "beep2 = replace_gtf_ids(gtf, m, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "b60f9021-d8ab-403e-a73f-3c424caf0772",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>ISM_subtype</th>\n",
       "      <th>adrenal_2mo_f_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>32839</td>\n",
       "      <td>86107</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 2, 1]</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx [1, 2, 1]</td>\n",
       "      <td>15</td>\n",
       "      <td>4281</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Prefix</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32839</td>\n",
       "      <td>86108</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 3, 2]</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx [1, 3, 2]</td>\n",
       "      <td>15</td>\n",
       "      <td>3779</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>32839</td>\n",
       "      <td>86111,86111</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>2</td>\n",
       "      <td>600</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32839</td>\n",
       "      <td>86109</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUSG00000020368.14 [3, 5, 4]</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx [3, 5, 4]</td>\n",
       "      <td>5</td>\n",
       "      <td>708</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32839</td>\n",
       "      <td>86110</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUSG00000020368.14 [4, 4, 5]</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx [4, 4, 5]</td>\n",
       "      <td>2</td>\n",
       "      <td>234</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_ID transcript_ID          annot_gene_id  \\\n",
       "0    32839         86107  ENSMUSG00000020368.15   \n",
       "1    32839         86108  ENSMUSG00000020368.15   \n",
       "2    32839   86111,86111  ENSMUSG00000020368.15   \n",
       "3    32839         86109  ENSMUSG00000020368.15   \n",
       "4    32839         86110  ENSMUSG00000020368.15   \n",
       "\n",
       "               annot_transcript_id annot_gene_name annot_transcript_name  \\\n",
       "0  ENSMUSG00000020368.14 [1, 2, 1]            Canx        Canx [1, 2, 1]   \n",
       "1  ENSMUSG00000020368.14 [1, 3, 2]            Canx        Canx [1, 3, 2]   \n",
       "2  ENSMUSG00000020368.14 [2, 1, 3]            Canx        Canx [2, 1, 3]   \n",
       "3  ENSMUSG00000020368.14 [3, 5, 4]            Canx        Canx [3, 5, 4]   \n",
       "4  ENSMUSG00000020368.14 [4, 4, 5]            Canx        Canx [4, 4, 5]   \n",
       "\n",
       "   n_exons  length gene_novelty transcript_novelty ISM_subtype  \\\n",
       "0       15    4281        Known              Known      Prefix   \n",
       "1       15    3779        Known              Known         NaN   \n",
       "2        2     600        Known              Known         NaN   \n",
       "3        5     708        Known              Known         NaN   \n",
       "4        2     234        Known              Known         NaN   \n",
       "\n",
       "   adrenal_2mo_f_1  \n",
       "0                8  \n",
       "1              371  \n",
       "2                2  \n",
       "3                1  \n",
       "4                0  "
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2d32db35-897f-4b5f-ae79-061a32aadd2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = 'tests/files/Canx_abundance.tsv'\n",
    "gtf = 'tests/files/Canx.gtf'\n",
    "m = 'tests/files/Canx_triplet_tid_map.tsv'\n",
    "h5 = '' # option to use / save in previous step as h5df?\n",
    "\n",
    "agg = True\n",
    "\n",
    "\n",
    "df = pd.read_csv(ab, sep='\\t')\n",
    "m_df = pd.read_csv(m, sep='\\t')\n",
    "\n",
    "# fix transcript ids in abundance file\n",
    "ab_map = m_df[['original_transcript_id', 'original_transcript_name', 'transcript_name', 'transcript_id']]\n",
    "df = df.merge(ab_map, left_on='annot_transcript_id', right_on='original_transcript_id')\n",
    "\n",
    "df.drop(['original_transcript_id', 'original_transcript_name', \n",
    "         'annot_transcript_id', 'annot_transcript_name'],\n",
    "        axis=1, inplace=True)\n",
    "df.rename({'transcript_name': 'annot_transcript_name',\n",
    "           'transcript_id': 'annot_transcript_id'},\n",
    "          axis=1, inplace=True)\n",
    "\n",
    "# aggregate counts if requested\n",
    "if agg:\n",
    "    gb_cols = ['gene_ID', 'annot_gene_id', 'annot_gene_name',\n",
    "               'gene_novelty', 'annot_transcript_name',\n",
    "               'annot_transcript_id']\n",
    "    \n",
    "    # handle properties which won't always correspond across the transcripts\n",
    "    # these are all subject to change\n",
    "    nov_rank, rank_nov = get_nov_ranks()\n",
    "    df['nov_rank'] = df.transcript_novelty.map(nov_rank)\n",
    "    df.drop('transcript_novelty', axis=1, inplace=True)\n",
    "    \n",
    "    ism_rank, rank_ism = get_ism_ranks()\n",
    "    df['ism_rank'] = df.ISM_subtype.map(ism_rank)\n",
    "    df.drop('ISM_subtype', axis=1, inplace=True)\n",
    "    \n",
    "    df['transcript_ID'] = df.transcript_ID.astype(str)\n",
    "    agg_dict = {'transcript_ID': ','.join,\n",
    "                'n_exons': 'mean', \n",
    "                'length': 'mean',\n",
    "                'nov_rank': 'min',\n",
    "                'ism_rank': 'min'}\n",
    "    cols = gb_cols + list(agg_dict.keys())\n",
    "    for c in list(set(df.columns)-set(cols)):\n",
    "        agg_dict[c] = 'sum'\n",
    "    \n",
    "    df = df.groupby(gb_cols).agg(agg_dict).reset_index()\n",
    "    df['transcript_novelty'] = df.nov_rank.map(rank_nov)\n",
    "    df.drop(['nov_rank'], axis=1, inplace=True)\n",
    "    df['ISM_subtype'] = df.ism_rank.map(rank_ism)\n",
    "    df.drop(['ism_rank'], axis=1, inplace=True)\n",
    "    \n",
    "# reorder columns\n",
    "c1 = get_non_dataset_cols()\n",
    "c2 = get_dataset_cols(df)\n",
    "df = df[c1+c2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "929512d7-d20a-4595-9f87-8f0ff8fe4111",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.read_gtf(gtf).df\n",
    "m_df = pd.read_csv(m, sep='\\t')\n",
    "\n",
    "# groupby transcripts that are the same\n",
    "gb_cols = ['gene_name', 'gene_id', 'transcript_triplet',\n",
    "           'transcript_id', 'transcript_name']\n",
    "temp = m_df[['transcript_id',\n",
    "             'original_transcript_id',\n",
    "             'original_transcript_name']].copy(deep=True)\n",
    "m_df = m_df.groupby(gb_cols).agg({'original_transcript_id': ','.join,\n",
    "                                  'original_transcript_name': ','.join}).reset_index()\n",
    "m_df = m_df.merge(temp, on='transcript_id', suffixes=('','_merge'))\n",
    "m_df.drop(['gene_name', 'gene_id', 'transcript_triplet'],\n",
    "          axis=1, inplace=True)\n",
    "\n",
    "# add new transcript ids \n",
    "df = df.merge(m_df, left_on=['transcript_id', 'transcript_name'],\n",
    "              right_on=['original_transcript_id_merge',\n",
    "                        'original_transcript_name_merge'],\n",
    "             suffixes=('_x', ''))\n",
    "\n",
    "# drop old tids\n",
    "df.drop(['transcript_id_x', 'transcript_name_x',\n",
    "         'original_transcript_name_merge'],\n",
    "        axis=1, inplace=True)\n",
    "\n",
    "# remove duplicated transcripts; just keeping the first one\n",
    "if agg:\n",
    "    temp = df[['transcript_id', 'original_transcript_id_merge']].drop_duplicates()\n",
    "    dupe_old_tids = temp.loc[temp.transcript_id.duplicated(keep='first'), 'original_transcript_id_merge']\n",
    "    df = df.loc[~df.original_transcript_id_merge.isin(dupe_old_tids)]  \n",
    "\n",
    "# drop last column\n",
    "df.drop('original_transcript_id_merge', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "96b424a8-2da2-4803-8fff-ce127d69ffdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = df[['transcript_id', 'original_transcript_id_merge']].drop_duplicates()\n",
    "dupe_old_tids = temp.loc[temp.transcript_id.duplicated(keep='first'), 'original_transcript_id_merge']\n",
    "temp = df.loc[~df.original_transcript_id_merge.isin(dupe_old_tids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "4ee3d044-1a37-44c9-bf69-c659aa33c833",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>original_transcript_id_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300896</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "79      chr11  HAVANA  transcript  50298837  50300990     .      -     .   \n",
       "80      chr11  HAVANA        exon  50300897  50300990     .      -     .   \n",
       "81      chr11  HAVANA         CDS  50300897  50300990     .      -     .   \n",
       "82      chr11  HAVANA        exon  50298837  50299344     .      -     .   \n",
       "83      chr11  HAVANA         CDS  50300896  50300990     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ... transcript_support_level  \\\n",
       "79  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "80  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "81  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "82  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "83  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "\n",
       "   ccdsid exon_number               exon_id     havana_transcript  \\\n",
       "79    NaN         NaN                   NaN  OTTMUST00000012413.1   \n",
       "80    NaN           1  ENSMUSE00000776476.1  OTTMUST00000012413.1   \n",
       "81    NaN           1  ENSMUSE00000776476.1  OTTMUST00000012413.1   \n",
       "82    NaN           2  ENSMUSE00000790569.1  OTTMUST00000012413.1   \n",
       "83    NaN           1  ENSMUSE00000776476.1  OTTMUST00000012413.1   \n",
       "\n",
       "                      transcript_id transcript_name  \\\n",
       "79  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "80  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "81  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "82  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "83  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "\n",
       "                       original_transcript_id original_transcript_name  \\\n",
       "79  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "80  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "81  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "82  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "83  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "\n",
       "   original_transcript_id_merge  \n",
       "79         ENSMUST00000155801.1  \n",
       "80         ENSMUST00000155801.1  \n",
       "81         ENSMUST00000155801.1  \n",
       "82         ENSMUST00000155801.1  \n",
       "83         ENSMUST00000155801.1  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.transcript_id == 'ENSMUSG00000020368.14 [2, 1, 3]']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "83507e1c-d606-4377-ac7c-c43422fb8a8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the start and stop coords for transcripts that are duplicated\n",
    "temp = df.loc[df.Feature == 'transcript'].copy(deep=True)\n",
    "df.head()\n",
    "cols = ['Start', 'End',\n",
    "        'transcript_id', 'Strand']\n",
    "temp = temp[cols].melt(id_vars=['transcript_id', 'Strand'], \n",
    "                       value_vars=['Start', 'End'],\n",
    "                       value_name='coord')\n",
    "temp.drop('variable', axis=1, inplace=True)\n",
    "rev_temp = temp.loc[temp.Strand == '-']\n",
    "fwd_temp = temp.loc[temp.Strand == '+']\n",
    "\n",
    "for strand, t_df in zip(['+', '-'], [fwd_temp, rev_temp]):\n",
    "    \n",
    "    # get min and max coords for each transcript\n",
    "    temp = t_df.copy(deep=True)\n",
    "    temp = temp.groupby('transcript_id').agg({'coord': ['min', 'max']})\n",
    "    temp = temp.droplevel(level=0, axis=1)\n",
    "    temp.reset_index(inplace=True)\n",
    "    \n",
    "    # these will be the new start / end coordinates for transcripts\n",
    "    # and their first exons (and possibly CDSs/UTRs/etc)\n",
    "    # if strand == '+':\n",
    "    d = {'min': 'new_Start', 'max': 'new_End'}\n",
    "    # elif strand == '-':\n",
    "    #     d = {'min': 'new_End', 'max': 'new_Start'}\n",
    "    \n",
    "    temp.rename(d, axis=1, inplace=True)\n",
    "    new_ends = temp.copy(deep=True)\n",
    "    \n",
    "    # get starts / ends from each transcript\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "42ff6105-522d-4ecd-bd13-b79a4bb840f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>new_Start</th>\n",
       "      <th>new_End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000020368.14 [1, 2, 1]</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000020368.14 [1, 3, 2]</td>\n",
       "      <td>50294466</td>\n",
       "      <td>50325673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>50298836</td>\n",
       "      <td>50300990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000020368.14 [3, 5, 4]</td>\n",
       "      <td>50296198</td>\n",
       "      <td>50299378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000020368.14 [4, 4, 5]</td>\n",
       "      <td>50297363</td>\n",
       "      <td>50298319</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     transcript_id  new_Start   new_End\n",
       "0  ENSMUSG00000020368.14 [1, 2, 1]   50293960  50325673\n",
       "1  ENSMUSG00000020368.14 [1, 3, 2]   50294466  50325673\n",
       "2  ENSMUSG00000020368.14 [2, 1, 3]   50298836  50300990\n",
       "3  ENSMUSG00000020368.14 [3, 5, 4]   50296198  50299378\n",
       "4  ENSMUSG00000020368.14 [4, 4, 5]   50297363  50298319"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_ends.head()\n",
    "# t_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "aba9a34d-ce56-494c-afe1-d3651ac4086b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp = df.merge(new_ends, how='left', on='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "355cc1f1-1343-4971-9ff6-afffeee6b95d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp.loc[temp.transcript_id == 'ENSMUSG00000020368.14 [2, 1, 3]', ['transcript_id', 'Start', 'End', 'new_Start', 'new_End']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "fa9fa80f-2ff1-4e37-b20b-d8eef3f0a632",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp['adj_Start'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "e42782c5-12b7-4406-a9ea-ed14ee6e7ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get list of all start / end coords for each transcript\n",
    "temp = df.loc[df.Feature == 'transcript'].copy(deep=True)\n",
    "df.head()\n",
    "cols = ['Start', 'End',\n",
    "        'transcript_id', 'Strand']\n",
    "temp = temp[cols]\n",
    "\n",
    "fwd_temp = temp.loc[temp.Strand == '+'].copy(deep=True)\n",
    "rev_temp = temp.loc[temp.Strand == '-'].copy(deep=True)\n",
    "\n",
    "for strand, t_df in zip(['+', '-'], [fwd_temp, rev_temp]):\n",
    "    if strand == '+':\n",
    "        t_df['old_Start'] = t_df[['Start', 'End']].min(axis=1)\n",
    "        t_df['old_End'] = t_df[['Start', 'End']].max(axis=1)\n",
    "    elif strand == '-':\n",
    "        t_df['old_Start'] = t_df[['Start', 'End']].max(axis=1)\n",
    "        t_df['old_End'] = t_df[['Start', 'End']].min(axis=1)\n",
    "    t_df.drop(['Start', 'End'], axis=1, inplace=True)\n",
    "    t_df.drop('Strand', axis=1, inplace=True)\n",
    "    \n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "cef24f8d-ea32-41d4-99e2-0d8955b602ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>original_transcript_id_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300896</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298836</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300896</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298836</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "      <td>ENSMUST00000155801.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "79      chr11  HAVANA  transcript  50298837  50300990     .      -     .   \n",
       "80      chr11  HAVANA        exon  50300897  50300990     .      -     .   \n",
       "81      chr11  HAVANA         CDS  50300897  50300990     .      -     .   \n",
       "82      chr11  HAVANA        exon  50298837  50299344     .      -     .   \n",
       "83      chr11  HAVANA         CDS  50300896  50300990     .      -     .   \n",
       "84      chr11  HAVANA  transcript  50298836  50300990     .      -     .   \n",
       "85      chr11  HAVANA        exon  50300896  50300990     .      -     .   \n",
       "86      chr11  HAVANA        exon  50298836  50299344     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ... transcript_support_level  \\\n",
       "79  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "80  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "81  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "82  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "83  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "84  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "85  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "86  ENSMUSG00000020368.14  protein_coding  ...                        3   \n",
       "\n",
       "   ccdsid exon_number               exon_id     havana_transcript  \\\n",
       "79    NaN         NaN                   NaN  OTTMUST00000012413.1   \n",
       "80    NaN           1  ENSMUSE00000776476.1  OTTMUST00000012413.1   \n",
       "81    NaN           1  ENSMUSE00000776476.1  OTTMUST00000012413.1   \n",
       "82    NaN           2  ENSMUSE00000790569.1  OTTMUST00000012413.1   \n",
       "83    NaN           1  ENSMUSE00000776476.1  OTTMUST00000012413.1   \n",
       "84    NaN         NaN                   NaN  OTTMUST00000012413.1   \n",
       "85    NaN           1  ENSMUSE00000776476.1  OTTMUST00000012413.1   \n",
       "86    NaN           2  ENSMUSE00000790569.1  OTTMUST00000012413.1   \n",
       "\n",
       "                      transcript_id transcript_name  \\\n",
       "79  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "80  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "81  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "82  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "83  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "84  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "85  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "86  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "\n",
       "                       original_transcript_id original_transcript_name  \\\n",
       "79  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "80  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "81  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "82  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "83  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "84  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "85  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "86  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004   \n",
       "\n",
       "   original_transcript_id_merge  \n",
       "79         ENSMUST00000155801.1  \n",
       "80         ENSMUST00000155801.1  \n",
       "81         ENSMUST00000155801.1  \n",
       "82         ENSMUST00000155801.1  \n",
       "83         ENSMUST00000155801.1  \n",
       "84         ENSMUST00000155801.2  \n",
       "85         ENSMUST00000155801.2  \n",
       "86         ENSMUST00000155801.2  \n",
       "\n",
       "[8 rows x 27 columns]"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id == 'ENSMUSG00000020368.14 [2, 1, 3]']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "d2ce0846-86d1-4bdf-bb77-d5ddfacdcd65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298116</td>\n",
       "      <td>50298319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000825549.1</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [4, 4, 5]</td>\n",
       "      <td>Canx [4, 4, 5]</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>Canx-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50297363</td>\n",
       "      <td>50297394</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000778008.1</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [4, 4, 5]</td>\n",
       "      <td>Canx [4, 4, 5]</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>Canx-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298836</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300896</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298836</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "77      chr11  HAVANA        exon  50298116  50298319     .      -     .   \n",
       "78      chr11  HAVANA        exon  50297363  50297394     .      -     .   \n",
       "79      chr11  HAVANA  transcript  50298837  50300990     .      -     .   \n",
       "80      chr11  HAVANA        exon  50300897  50300990     .      -     .   \n",
       "81      chr11  HAVANA         CDS  50300897  50300990     .      -     .   \n",
       "82      chr11  HAVANA        exon  50298837  50299344     .      -     .   \n",
       "83      chr11  HAVANA         CDS  50300897  50300990     .      -     .   \n",
       "84      chr11  HAVANA  transcript  50298836  50300990     .      -     .   \n",
       "85      chr11  HAVANA        exon  50300896  50300990     .      -     .   \n",
       "86      chr11  HAVANA        exon  50298836  50299344     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ...                 tag  \\\n",
       "77  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "78  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "79  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "80  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "81  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "82  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "83  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "84  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
       "85  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
       "86  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
       "\n",
       "   transcript_support_level ccdsid exon_number               exon_id  \\\n",
       "77                        5    NaN           1  ENSMUSE00000825549.1   \n",
       "78                        5    NaN           2  ENSMUSE00000778008.1   \n",
       "79                        3    NaN         NaN                   NaN   \n",
       "80                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "81                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "82                        3    NaN           2  ENSMUSE00000790569.1   \n",
       "83                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "84                        3    NaN         NaN                   NaN   \n",
       "85                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "86                        3    NaN           2  ENSMUSE00000790569.1   \n",
       "\n",
       "       havana_transcript                    transcript_id transcript_name  \\\n",
       "77  OTTMUST00000012411.1  ENSMUSG00000020368.14 [4, 4, 5]  Canx [4, 4, 5]   \n",
       "78  OTTMUST00000012411.1  ENSMUSG00000020368.14 [4, 4, 5]  Canx [4, 4, 5]   \n",
       "79  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "80  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "81  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "82  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "83  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "84  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "85  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "86  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "\n",
       "                       original_transcript_id original_transcript_name  \n",
       "77                       ENSMUST00000153068.1                 Canx-002  \n",
       "78                       ENSMUST00000153068.1                 Canx-002  \n",
       "79  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "80  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "81  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "82  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "83  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "84  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "85  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "86  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "\n",
       "[10 rows x 26 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "87d32206-3cad-40c1-8f0f-67498503ac09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp['min'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "411226d7-ff95-4768-9f03-aaf2461f3b39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298116</td>\n",
       "      <td>50298319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000825549.1</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [4, 4, 5]</td>\n",
       "      <td>Canx [4, 4, 5]</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>Canx-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50297363</td>\n",
       "      <td>50297394</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000778008.1</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [4, 4, 5]</td>\n",
       "      <td>Canx [4, 4, 5]</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>Canx-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298836</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300896</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298836</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "77      chr11  HAVANA        exon  50298116  50298319     .      -     .   \n",
       "78      chr11  HAVANA        exon  50297363  50297394     .      -     .   \n",
       "79      chr11  HAVANA  transcript  50298837  50300990     .      -     .   \n",
       "80      chr11  HAVANA        exon  50300897  50300990     .      -     .   \n",
       "81      chr11  HAVANA         CDS  50300897  50300990     .      -     .   \n",
       "82      chr11  HAVANA        exon  50298837  50299344     .      -     .   \n",
       "83      chr11  HAVANA         CDS  50300897  50300990     .      -     .   \n",
       "84      chr11  HAVANA  transcript  50298836  50300990     .      -     .   \n",
       "85      chr11  HAVANA        exon  50300896  50300990     .      -     .   \n",
       "86      chr11  HAVANA        exon  50298836  50299344     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ...                 tag  \\\n",
       "77  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "78  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "79  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "80  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "81  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "82  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "83  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "84  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
       "85  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
       "86  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
       "\n",
       "   transcript_support_level ccdsid exon_number               exon_id  \\\n",
       "77                        5    NaN           1  ENSMUSE00000825549.1   \n",
       "78                        5    NaN           2  ENSMUSE00000778008.1   \n",
       "79                        3    NaN         NaN                   NaN   \n",
       "80                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "81                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "82                        3    NaN           2  ENSMUSE00000790569.1   \n",
       "83                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "84                        3    NaN         NaN                   NaN   \n",
       "85                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "86                        3    NaN           2  ENSMUSE00000790569.1   \n",
       "\n",
       "       havana_transcript                    transcript_id transcript_name  \\\n",
       "77  OTTMUST00000012411.1  ENSMUSG00000020368.14 [4, 4, 5]  Canx [4, 4, 5]   \n",
       "78  OTTMUST00000012411.1  ENSMUSG00000020368.14 [4, 4, 5]  Canx [4, 4, 5]   \n",
       "79  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "80  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "81  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "82  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "83  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "84  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "85  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "86  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "\n",
       "                       original_transcript_id original_transcript_name  \n",
       "77                       ENSMUST00000153068.1                 Canx-002  \n",
       "78                       ENSMUST00000153068.1                 Canx-002  \n",
       "79  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "80  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "81  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "82  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "83  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "84  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "85  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "86  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "\n",
       "[10 rows x 26 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 639,
   "id": "658726ba-29ef-43f5-b01e-f30bfe68e212",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50297273</td>\n",
       "      <td>50297394</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00001293345.1</td>\n",
       "      <td>OTTMUST00000012412.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [3, 5, 4]</td>\n",
       "      <td>Canx [3, 5, 4]</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>Canx-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50297057</td>\n",
       "      <td>50297137</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSE00001310416.1</td>\n",
       "      <td>OTTMUST00000012412.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [3, 5, 4]</td>\n",
       "      <td>Canx [3, 5, 4]</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>Canx-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50296198</td>\n",
       "      <td>50296335</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSMUSE00000827057.1</td>\n",
       "      <td>OTTMUST00000012412.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [3, 5, 4]</td>\n",
       "      <td>Canx [3, 5, 4]</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>Canx-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50297363</td>\n",
       "      <td>50298319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [4, 4, 5]</td>\n",
       "      <td>Canx [4, 4, 5]</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>Canx-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298116</td>\n",
       "      <td>50298319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000825549.1</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [4, 4, 5]</td>\n",
       "      <td>Canx [4, 4, 5]</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>Canx-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50297363</td>\n",
       "      <td>50297394</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000778008.1</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [4, 4, 5]</td>\n",
       "      <td>Canx [4, 4, 5]</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>Canx-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "73      chr11  HAVANA        exon  50297273  50297394     .      -     .   \n",
       "74      chr11  HAVANA        exon  50297057  50297137     .      -     .   \n",
       "75      chr11  HAVANA        exon  50296198  50296335     .      -     .   \n",
       "76      chr11  HAVANA  transcript  50297363  50298319     .      -     .   \n",
       "77      chr11  HAVANA        exon  50298116  50298319     .      -     .   \n",
       "78      chr11  HAVANA        exon  50297363  50297394     .      -     .   \n",
       "79      chr11  HAVANA  transcript  50298837  50300990     .      -     .   \n",
       "80      chr11  HAVANA        exon  50300897  50300990     .      -     .   \n",
       "81      chr11  HAVANA         CDS  50300897  50300990     .      -     .   \n",
       "82      chr11  HAVANA        exon  50298837  50299344     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ...                 tag  \\\n",
       "73  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "74  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "75  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "76  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "77  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "78  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "79  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "80  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "81  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "82  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "\n",
       "   transcript_support_level ccdsid exon_number               exon_id  \\\n",
       "73                        2    NaN           3  ENSMUSE00001293345.1   \n",
       "74                        2    NaN           4  ENSMUSE00001310416.1   \n",
       "75                        2    NaN           5  ENSMUSE00000827057.1   \n",
       "76                        5    NaN         NaN                   NaN   \n",
       "77                        5    NaN           1  ENSMUSE00000825549.1   \n",
       "78                        5    NaN           2  ENSMUSE00000778008.1   \n",
       "79                        3    NaN         NaN                   NaN   \n",
       "80                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "81                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "82                        3    NaN           2  ENSMUSE00000790569.1   \n",
       "\n",
       "       havana_transcript                    transcript_id transcript_name  \\\n",
       "73  OTTMUST00000012412.1  ENSMUSG00000020368.14 [3, 5, 4]  Canx [3, 5, 4]   \n",
       "74  OTTMUST00000012412.1  ENSMUSG00000020368.14 [3, 5, 4]  Canx [3, 5, 4]   \n",
       "75  OTTMUST00000012412.1  ENSMUSG00000020368.14 [3, 5, 4]  Canx [3, 5, 4]   \n",
       "76  OTTMUST00000012411.1  ENSMUSG00000020368.14 [4, 4, 5]  Canx [4, 4, 5]   \n",
       "77  OTTMUST00000012411.1  ENSMUSG00000020368.14 [4, 4, 5]  Canx [4, 4, 5]   \n",
       "78  OTTMUST00000012411.1  ENSMUSG00000020368.14 [4, 4, 5]  Canx [4, 4, 5]   \n",
       "79  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "80  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "81  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "82  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "\n",
       "                       original_transcript_id original_transcript_name  \n",
       "73                       ENSMUST00000146979.1                 Canx-003  \n",
       "74                       ENSMUST00000146979.1                 Canx-003  \n",
       "75                       ENSMUST00000146979.1                 Canx-003  \n",
       "76                       ENSMUST00000153068.1                 Canx-002  \n",
       "77                       ENSMUST00000153068.1                 Canx-002  \n",
       "78                       ENSMUST00000153068.1                 Canx-002  \n",
       "79  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "80  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "81  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "82  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "\n",
       "[10 rows x 26 columns]"
      ]
     },
     "execution_count": 639,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# if we're only reporting one transcript per triplet\n",
    "if agg:\n",
    "    print(len(df.index))\n",
    "    df = df.loc[~(df.duplicated(subset=['transcript_id',\n",
    "                                         'transcript_name',\n",
    "                                         'exon_id',\n",
    "                                         'Feature'], keep='first'))]\n",
    "    print(len(df.index))\n",
    "    \n",
    "# drop original_tid_merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 624,
   "id": "126dd6a3-0f21-49da-90fa-4904e65f6666",
   "metadata": {},
   "outputs": [],
   "source": [
    "# is the deduplication really working correctly? What happens to duplicated cds / other features?\n",
    "df = pr.read_gtf('tests/files/Canx.gtf').df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 625,
   "id": "1dc7cab7-3f4c-463e-b7a9-b22c5eaac253",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gene', 'transcript', 'exon', 'CDS', 'start_codon', 'stop_codon', 'UTR']\n",
       "Categories (7, object): ['gene', 'transcript', 'exon', 'CDS', 'start_codon', 'stop_codon', 'UTR']"
      ]
     },
     "execution_count": 625,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.Feature.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 628,
   "id": "9e16221a-1aa2-40e5-9559-978d7dc5a53c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50325492</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000371309.6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311780</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>50311774</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50310371</td>\n",
       "      <td>50310430</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSE00000103866.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50310371</td>\n",
       "      <td>50310430</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSE00000103866.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50308787</td>\n",
       "      <td>50308929</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSMUSE00000296570.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50308787</td>\n",
       "      <td>50308929</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSMUSE00000296570.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50308309</td>\n",
       "      <td>50308391</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSMUSE00000103859.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50308309</td>\n",
       "      <td>50308391</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSMUSE00000103859.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50307469</td>\n",
       "      <td>50307662</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSMUSE00000103857.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50307469</td>\n",
       "      <td>50307662</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSMUSE00000103857.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50304325</td>\n",
       "      <td>50304515</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSMUSE00000103855.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50304325</td>\n",
       "      <td>50304515</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSMUSE00000103855.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50301757</td>\n",
       "      <td>50301871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSMUSE00000103853.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50301757</td>\n",
       "      <td>50301871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSMUSE00000103853.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50301054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>10</td>\n",
       "      <td>ENSMUSE00000103851.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50301054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>10</td>\n",
       "      <td>ENSMUSE00000103851.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50299128</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>11</td>\n",
       "      <td>ENSMUSE00000103856.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50299128</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>11</td>\n",
       "      <td>ENSMUSE00000103856.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298116</td>\n",
       "      <td>50298236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSMUSE00001261330.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50298116</td>\n",
       "      <td>50298236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSMUSE00001261330.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50297273</td>\n",
       "      <td>50297394</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>13</td>\n",
       "      <td>ENSMUSE00001288818.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50297273</td>\n",
       "      <td>50297394</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>13</td>\n",
       "      <td>ENSMUSE00001288818.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50297057</td>\n",
       "      <td>50297137</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>14</td>\n",
       "      <td>ENSMUSE00001256122.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50297057</td>\n",
       "      <td>50297137</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>14</td>\n",
       "      <td>ENSMUSE00001256122.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50296335</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSMUSE00001061703.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50296284</td>\n",
       "      <td>50296335</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSMUSE00001061703.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>50296281</td>\n",
       "      <td>50296284</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSMUSE00001061703.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>UTR</td>\n",
       "      <td>50325492</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000371309.6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>UTR</td>\n",
       "      <td>50311777</td>\n",
       "      <td>50311780</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>UTR</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50296284</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSMUSE00001061703.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome   Source      Feature     Start       End Score Strand Frame  \\\n",
       "1       chr11  ENSEMBL   transcript  50293960  50325673     .      -     .   \n",
       "2       chr11  ENSEMBL         exon  50325492  50325673     .      -     .   \n",
       "3       chr11  ENSEMBL         exon  50311603  50311780     .      -     .   \n",
       "4       chr11  ENSEMBL          CDS  50311603  50311777     .      -     0   \n",
       "5       chr11  ENSEMBL  start_codon  50311774  50311777     .      -     0   \n",
       "6       chr11  ENSEMBL         exon  50310793  50310867     .      -     .   \n",
       "7       chr11  ENSEMBL          CDS  50310793  50310867     .      -     0   \n",
       "8       chr11  ENSEMBL         exon  50310371  50310430     .      -     .   \n",
       "9       chr11  ENSEMBL          CDS  50310371  50310430     .      -     1   \n",
       "10      chr11  ENSEMBL         exon  50308787  50308929     .      -     .   \n",
       "11      chr11  ENSEMBL          CDS  50308787  50308929     .      -     2   \n",
       "12      chr11  ENSEMBL         exon  50308309  50308391     .      -     .   \n",
       "13      chr11  ENSEMBL          CDS  50308309  50308391     .      -     1   \n",
       "14      chr11  ENSEMBL         exon  50307469  50307662     .      -     .   \n",
       "15      chr11  ENSEMBL          CDS  50307469  50307662     .      -     0   \n",
       "16      chr11  ENSEMBL         exon  50304325  50304515     .      -     .   \n",
       "17      chr11  ENSEMBL          CDS  50304325  50304515     .      -     2   \n",
       "18      chr11  ENSEMBL         exon  50301757  50301871     .      -     .   \n",
       "19      chr11  ENSEMBL          CDS  50301757  50301871     .      -     1   \n",
       "20      chr11  ENSEMBL         exon  50300897  50301054     .      -     .   \n",
       "21      chr11  ENSEMBL          CDS  50300897  50301054     .      -     1   \n",
       "22      chr11  ENSEMBL         exon  50299128  50299344     .      -     .   \n",
       "23      chr11  ENSEMBL          CDS  50299128  50299344     .      -     0   \n",
       "24      chr11  ENSEMBL         exon  50298116  50298236     .      -     .   \n",
       "25      chr11  ENSEMBL          CDS  50298116  50298236     .      -     0   \n",
       "26      chr11  ENSEMBL         exon  50297273  50297394     .      -     .   \n",
       "27      chr11  ENSEMBL          CDS  50297273  50297394     .      -     0   \n",
       "28      chr11  ENSEMBL         exon  50297057  50297137     .      -     .   \n",
       "29      chr11  ENSEMBL          CDS  50297057  50297137     .      -     2   \n",
       "30      chr11  ENSEMBL         exon  50293960  50296335     .      -     .   \n",
       "31      chr11  ENSEMBL          CDS  50296284  50296335     .      -     0   \n",
       "32      chr11  ENSEMBL   stop_codon  50296281  50296284     .      -     0   \n",
       "33      chr11  ENSEMBL          UTR  50325492  50325673     .      -     .   \n",
       "34      chr11  ENSEMBL          UTR  50311777  50311780     .      -     .   \n",
       "35      chr11  ENSEMBL          UTR  50293960  50296284     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ... transcript_type  \\\n",
       "1   ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "2   ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "3   ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "4   ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "5   ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "6   ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "7   ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "8   ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "9   ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "10  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "11  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "12  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "13  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "14  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "15  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "16  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "17  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "18  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "19  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "20  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "21  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "22  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "23  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "24  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "25  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "26  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "27  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "28  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "29  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "30  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "31  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "32  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "33  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "34  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "35  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "\n",
       "   transcript_status transcript_name            protein_id   tag  \\\n",
       "1              KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "2              KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "3              KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "4              KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "5              KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "6              KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "7              KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "8              KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "9              KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "10             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "11             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "12             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "13             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "14             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "15             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "16             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "17             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "18             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "19             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "20             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "21             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "22             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "23             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "24             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "25             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "26             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "27             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "28             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "29             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "30             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "31             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "32             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "33             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "34             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "35             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "\n",
       "   transcript_support_level       ccdsid exon_number               exon_id  \\\n",
       "1                         1  CCDS24633.1         NaN                   NaN   \n",
       "2                         1  CCDS24633.1           1  ENSMUSE00000371309.6   \n",
       "3                         1  CCDS24633.1           2  ENSMUSE00001016144.1   \n",
       "4                         1  CCDS24633.1           2  ENSMUSE00001016144.1   \n",
       "5                         1  CCDS24633.1           2  ENSMUSE00001016144.1   \n",
       "6                         1  CCDS24633.1           3  ENSMUSE00000103860.1   \n",
       "7                         1  CCDS24633.1           3  ENSMUSE00000103860.1   \n",
       "8                         1  CCDS24633.1           4  ENSMUSE00000103866.1   \n",
       "9                         1  CCDS24633.1           4  ENSMUSE00000103866.1   \n",
       "10                        1  CCDS24633.1           5  ENSMUSE00000296570.1   \n",
       "11                        1  CCDS24633.1           5  ENSMUSE00000296570.1   \n",
       "12                        1  CCDS24633.1           6  ENSMUSE00000103859.1   \n",
       "13                        1  CCDS24633.1           6  ENSMUSE00000103859.1   \n",
       "14                        1  CCDS24633.1           7  ENSMUSE00000103857.1   \n",
       "15                        1  CCDS24633.1           7  ENSMUSE00000103857.1   \n",
       "16                        1  CCDS24633.1           8  ENSMUSE00000103855.1   \n",
       "17                        1  CCDS24633.1           8  ENSMUSE00000103855.1   \n",
       "18                        1  CCDS24633.1           9  ENSMUSE00000103853.1   \n",
       "19                        1  CCDS24633.1           9  ENSMUSE00000103853.1   \n",
       "20                        1  CCDS24633.1          10  ENSMUSE00000103851.1   \n",
       "21                        1  CCDS24633.1          10  ENSMUSE00000103851.1   \n",
       "22                        1  CCDS24633.1          11  ENSMUSE00000103856.1   \n",
       "23                        1  CCDS24633.1          11  ENSMUSE00000103856.1   \n",
       "24                        1  CCDS24633.1          12  ENSMUSE00001261330.1   \n",
       "25                        1  CCDS24633.1          12  ENSMUSE00001261330.1   \n",
       "26                        1  CCDS24633.1          13  ENSMUSE00001288818.1   \n",
       "27                        1  CCDS24633.1          13  ENSMUSE00001288818.1   \n",
       "28                        1  CCDS24633.1          14  ENSMUSE00001256122.1   \n",
       "29                        1  CCDS24633.1          14  ENSMUSE00001256122.1   \n",
       "30                        1  CCDS24633.1          15  ENSMUSE00001061703.1   \n",
       "31                        1  CCDS24633.1          15  ENSMUSE00001061703.1   \n",
       "32                        1  CCDS24633.1          15  ENSMUSE00001061703.1   \n",
       "33                        1  CCDS24633.1           1  ENSMUSE00000371309.6   \n",
       "34                        1  CCDS24633.1           2  ENSMUSE00001016144.1   \n",
       "35                        1  CCDS24633.1          15  ENSMUSE00001061703.1   \n",
       "\n",
       "   havana_transcript  \n",
       "1                NaN  \n",
       "2                NaN  \n",
       "3                NaN  \n",
       "4                NaN  \n",
       "5                NaN  \n",
       "6                NaN  \n",
       "7                NaN  \n",
       "8                NaN  \n",
       "9                NaN  \n",
       "10               NaN  \n",
       "11               NaN  \n",
       "12               NaN  \n",
       "13               NaN  \n",
       "14               NaN  \n",
       "15               NaN  \n",
       "16               NaN  \n",
       "17               NaN  \n",
       "18               NaN  \n",
       "19               NaN  \n",
       "20               NaN  \n",
       "21               NaN  \n",
       "22               NaN  \n",
       "23               NaN  \n",
       "24               NaN  \n",
       "25               NaN  \n",
       "26               NaN  \n",
       "27               NaN  \n",
       "28               NaN  \n",
       "29               NaN  \n",
       "30               NaN  \n",
       "31               NaN  \n",
       "32               NaN  \n",
       "33               NaN  \n",
       "34               NaN  \n",
       "35               NaN  \n",
       "\n",
       "[35 rows x 24 columns]"
      ]
     },
     "execution_count": 628,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_name == 'Canx-201']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 630,
   "id": "48923dbc-0b0f-4192-b166-987d1692d84f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50325492</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000371309.6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311780</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>50311774</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50310371</td>\n",
       "      <td>50310430</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSE00000103866.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50310371</td>\n",
       "      <td>50310430</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSE00000103866.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Source      Feature     Start       End Score Strand Frame  \\\n",
       "0      chr11   HAVANA         gene  50293960  50325673     .      -     .   \n",
       "1      chr11  ENSEMBL   transcript  50293960  50325673     .      -     .   \n",
       "2      chr11  ENSEMBL         exon  50325492  50325673     .      -     .   \n",
       "3      chr11  ENSEMBL         exon  50311603  50311780     .      -     .   \n",
       "4      chr11  ENSEMBL          CDS  50311603  50311777     .      -     0   \n",
       "5      chr11  ENSEMBL  start_codon  50311774  50311777     .      -     0   \n",
       "6      chr11  ENSEMBL         exon  50310793  50310867     .      -     .   \n",
       "7      chr11  ENSEMBL          CDS  50310793  50310867     .      -     0   \n",
       "8      chr11  ENSEMBL         exon  50310371  50310430     .      -     .   \n",
       "9      chr11  ENSEMBL          CDS  50310371  50310430     .      -     1   \n",
       "\n",
       "                 gene_id       gene_type  ... transcript_type  \\\n",
       "0  ENSMUSG00000020368.14  protein_coding  ...             NaN   \n",
       "1  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "2  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "3  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "4  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "5  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "6  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "7  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "8  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "9  ENSMUSG00000020368.14  protein_coding  ...  protein_coding   \n",
       "\n",
       "  transcript_status transcript_name            protein_id   tag  \\\n",
       "0               NaN             NaN                   NaN   NaN   \n",
       "1             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "2             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "3             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "4             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "5             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "6             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "7             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "8             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "9             KNOWN        Canx-201  ENSMUSP00000137440.1  CCDS   \n",
       "\n",
       "  transcript_support_level       ccdsid exon_number               exon_id  \\\n",
       "0                      NaN          NaN         NaN                   NaN   \n",
       "1                        1  CCDS24633.1         NaN                   NaN   \n",
       "2                        1  CCDS24633.1           1  ENSMUSE00000371309.6   \n",
       "3                        1  CCDS24633.1           2  ENSMUSE00001016144.1   \n",
       "4                        1  CCDS24633.1           2  ENSMUSE00001016144.1   \n",
       "5                        1  CCDS24633.1           2  ENSMUSE00001016144.1   \n",
       "6                        1  CCDS24633.1           3  ENSMUSE00000103860.1   \n",
       "7                        1  CCDS24633.1           3  ENSMUSE00000103860.1   \n",
       "8                        1  CCDS24633.1           4  ENSMUSE00000103866.1   \n",
       "9                        1  CCDS24633.1           4  ENSMUSE00000103866.1   \n",
       "\n",
       "  havana_transcript  \n",
       "0               NaN  \n",
       "1               NaN  \n",
       "2               NaN  \n",
       "3               NaN  \n",
       "4               NaN  \n",
       "5               NaN  \n",
       "6               NaN  \n",
       "7               NaN  \n",
       "8               NaN  \n",
       "9               NaN  \n",
       "\n",
       "[10 rows x 24 columns]"
      ]
     },
     "execution_count": 630,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86b6f508-719b-4ed3-98a3-2d6da9d77e8c",
   "metadata": {},
   "source": [
    "## change add_triplets to make it not return gtf and return IC reference\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eafeb834-e1e0-4787-b921-ede23a34463d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transcript_ref(fname):\n",
    "    \"\"\"\n",
    "    Get transcripts with necessary tags to order based on basic set etc.\n",
    "\n",
    "    Parameters:\n",
    "        fname (str): Path to GTF file\n",
    "\n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame of only transcript entries\n",
    "            with tags parsed out\n",
    "    \"\"\"\n",
    "    df = pr.read_gtf(fname, duplicate_attr=True).df\n",
    "    df = df.loc[df.Feature == 'transcript']\n",
    "    df['basic_set'] = df.tag.str.contains('basic')\n",
    "    df['temp'] = df.tag.str.split('appris_principal_', n=1, expand=True)[1]\n",
    "    df['appris_principal'] = df.temp.str.split(',', n=1, expand=True)[0]\n",
    "\n",
    "    return df\n",
    "\n",
    "def number_tss_ic_tes(df, mode):\n",
    "    \"\"\"\n",
    "    Assign a number to each tss, intron chain, or tes per gene based on the\n",
    "    status of the transcript\n",
    "\n",
    "    Parameters:\n",
    "        df (pandas DataFrame): DataFrame derived from PyRanges\n",
    "            with gene id, tss/ic/tes id, and tags to order transcripts\n",
    "            with\n",
    "        mode (str): {'tss', 'ic', 'tes'}\n",
    "\n",
    "    Returns:\n",
    "        df (pandas DataFrame): DataFrame with an identifier for each unique\n",
    "            tss/ic/tes per gene\n",
    "    \"\"\"\n",
    "    # groupby feature but record which feature\n",
    "    # each transcript id uses\n",
    "    cols = ['transcript_id', 'Chromosome', 'Strand',\n",
    "            mode, 'basic_set', 'appris_principal', 'gene_id']\n",
    "    df = df[cols].groupby(['Chromosome', 'Strand', \n",
    "                           mode, 'gene_id']).agg({'transcript_id': ','.join,\n",
    "                                     'basic_set': 'max',\n",
    "                                     'appris_principal': 'min'}).reset_index()\n",
    "\n",
    "    # compute intron chain number\n",
    "    df['{}_num'.format(mode)] = df.sort_values(by=['gene_id', 'appris_principal', 'basic_set'],\n",
    "                                 ascending=[True, False, True],\n",
    "                                 na_position='last')\\\n",
    "                                 .groupby(['gene_id'])\\\n",
    "                                 .cumcount() + 1\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57f02f13-b811-411c-b8b4-4df78380b010",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3879450b-387e-4ab7-8f36-85e5f8ded73d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def add_triplets(gtf, tss_bed, tes_bed):\n",
    "#     \"\"\"\n",
    "#     Merges end regions with ends from GTF and assigns triplets to each detected\n",
    "#     intron chain, tss, and tes for each gene.\n",
    "#     Preferentially orders based on transcript support level (appris_principal)\n",
    "#     and which transcripts are in the basic annotation\n",
    "\n",
    "#     Parameters:\n",
    "#         gtf (str): File path to GTF\n",
    "#         tss_bed (str): File path to TSS bed; output from `aggregate_ends` or\n",
    "#             `get_ends_from_gtf`\n",
    "#         tes_bed (str): File path to TES bed: output from `aggregate_ends` or\n",
    "#             `get_ends_from_gtf`\n",
    "\n",
    "#     Returns:\n",
    "#         gtf (pyranges PyRanges): GTF PyRanges object with added field\n",
    "#             \"transcript_triplet\"\n",
    "#         tss_bed (pyranges PyRanges): BED PyRanges object with TSS named\n",
    "#             gene_id_tss_#\n",
    "#         tes_bed (pyranges PyRanges): BED PyRanges object with TES named\n",
    "#             gene_id_tes_#\n",
    "#         df (pandas DataFrame): DataFrame including new and old transcript IDs\n",
    "#     \"\"\"\n",
    "\n",
    "gtf = 'tests/files/Canx.gtf'\n",
    "tss_bed ='tests/files/Canx_tss.bed'\n",
    "tes_bed = 'tests/files/Canx_tes.bed'\n",
    "\n",
    "### intron chain annotation ###\n",
    "\n",
    "# get basic status and appris_principal tag for each transcript\n",
    "t_df = get_transcript_ref(gtf)\n",
    "t_df = t_df[['transcript_id', 'gene_id', 'basic_set', 'appris_principal']]\n",
    "\n",
    "# get unique intron chains from gtf\n",
    "df = pr.read_gtf(gtf).df\n",
    "\n",
    "# restrict to exon entries\n",
    "df = df.loc[df.Feature == 'exon']\n",
    "cols = ['Chromosome', 'Strand', 'Start', 'End', 'transcript_id', 'gene_id']\n",
    "df = df[cols]\n",
    "\n",
    "# melt to isolate individual coordinates\n",
    "df = pd.melt(df, id_vars=['Chromosome', 'Strand', 'transcript_id', 'gene_id'],\n",
    "            value_vars=['Start', 'End'],\n",
    "            value_name='Coord')\n",
    "df.drop('variable', axis=1, inplace=True)\n",
    "\n",
    "# sort to order coordinates correctly\n",
    "df.Coord = df.Coord.astype(int)\n",
    "fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "\n",
    "fwd.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                ascending=[True, True, True])\n",
    "rev.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                ascending=[True, True, False])\n",
    "df = pd.concat([fwd, rev])\n",
    "\n",
    "# create intron chain strings\n",
    "df.Coord = df.Coord.astype(str)\n",
    "df = df.groupby(['Chromosome', 'Strand', 'transcript_id'])['Coord'].apply('-'.join).reset_index()\n",
    "\n",
    "# remove tss and tes from intron chain\n",
    "df['temp'] = df.Coord.str.split('-', n=1, expand=True)[1]\n",
    "df['ic'] = df.temp.str.rsplit('-', n=1, expand=True)[0]\n",
    "\n",
    "# add basic annotation, appris principal number, and gene id\n",
    "df = df.merge(t_df, on='transcript_id', how='left')\n",
    "\n",
    "# add number for each unique intron chain\n",
    "df = number_tss_ic_tes(df, mode='ic')\n",
    "\n",
    "# make coords into tuple\n",
    "df['ic'] = df.ic.str.split('-')\n",
    "df['ic'] = [tuple(c) for c in df.ic.tolist()]\n",
    "ic = df.copy(deep=True)\n",
    "ic.rename({'ic': 'coordinates',\n",
    "           'Chromosome': 'chrom',\n",
    "           'Strand': 'strand'},\n",
    "           axis=1, inplace=True)\n",
    "ic['ic_id'] = ic['gene_id']+'_'+ic.ic_num.astype(str)\n",
    "ic.drop(['basic_set', 'appris_principal',\n",
    "         'gene_id', 'ic_num'],\n",
    "        axis=1, inplace=True)\n",
    "\n",
    "\n",
    "# keep track of which transcripts use each intron chain\n",
    "df.transcript_id = df.transcript_id.str.split(',')\n",
    "ttrip_df = df[['transcript_id', 'ic_num']].explode(column='transcript_id')\n",
    "\n",
    "### tss / tes numbering ###\n",
    "beds = {}\n",
    "for mode, bed_file in zip(['tss', 'tes'], [tss_bed, tes_bed]):\n",
    "\n",
    "    ends = pr.read_bed(bed_file).df\n",
    "    ends.rename({'ThickStart': 'gene_id'}, axis=1, inplace=True)\n",
    "    ends[mode] = [i for i in range(len(ends.index))]\n",
    "    ends = pr.PyRanges(ends)\n",
    "\n",
    "    # merge gtf ends with bed ends\n",
    "    df = pr.read_gtf(gtf)\n",
    "    if mode == 'tss':\n",
    "        df = df.features.tss().df\n",
    "    elif mode == 'tes':\n",
    "        df = df.features.tes().df\n",
    "\n",
    "    # merge in information about transcript status\n",
    "    t_df = get_transcript_ref(gtf)\n",
    "    t_df = t_df[['transcript_id', 'appris_principal', 'basic_set']]\n",
    "    df = df.merge(t_df, how='left', on='transcript_id')\n",
    "    df = pr.PyRanges(df)\n",
    "\n",
    "    # join with bed ends\n",
    "    df = df.join(ends,\n",
    "                 strandedness='same',\n",
    "                 how='left').df\n",
    "    df = df.loc[df.gene_id == df.gene_id_b]\n",
    "\n",
    "    # add number for each unique end region\n",
    "    df = number_tss_ic_tes(df, mode=mode)\n",
    "\n",
    "    # keep track of which transcripts use each intron chain\n",
    "    df.transcript_id = df.transcript_id.str.split(',')\n",
    "    temp = df[['transcript_id', '{}_num'.format(mode)]].explode(column='transcript_id')\n",
    "    ttrip_df = ttrip_df.merge(temp, on='transcript_id')\n",
    "\n",
    "    # add numbers back into bed file\n",
    "    ends = ends.df\n",
    "    ends = ends.merge(df[[mode, '{}_num'.format(mode)]], on=mode)\n",
    "\n",
    "    ends['Name'] = ends.gene_id+'_'+ends['{}_num'.format(mode)].astype(str)\n",
    "    ends.drop([mode, 'gene_id', '{}_num'.format(mode)], axis=1, inplace=True)\n",
    "    beds[mode] = pr.PyRanges(ends)\n",
    "\n",
    "# create formatted string to represent each transcript\n",
    "ttrip_df['transcript_triplet'] = '['+ttrip_df['tss_num'].astype(str)+', '+\\\n",
    "                                     ttrip_df['ic_num'].astype(str)+', '+\\\n",
    "                                     ttrip_df['tes_num'].astype(str)+']'\n",
    "\n",
    "# final gtf file formatting\n",
    "df = pr.read_gtf(gtf, duplicate_attr=True).df\n",
    "df = df.merge(ttrip_df[['transcript_id', 'transcript_triplet']],\n",
    "              how='left',\n",
    "              on='transcript_id')\n",
    "gtf = pr.PyRanges(df)\n",
    "\n",
    "# transcript to triplet reference file\n",
    "\n",
    "# add gene name from transcript name\n",
    "df = df.loc[df.Feature == 'transcript']\n",
    "if 'gene_name' not in df.columns:\n",
    "    df['gene_name'] = df.transcript_name.str.split('-', n=1, expand=True)[0]\n",
    "df = df[['transcript_id', 'transcript_name',\n",
    "         'gene_name', 'gene_id',\n",
    "         'transcript_triplet']]\n",
    "df.rename({'transcript_id': 'original_transcript_id',\n",
    "           'transcript_name': 'original_transcript_name'},\n",
    "           axis=1, inplace=True)\n",
    "df['transcript_id'] = df.gene_id+' '+df.transcript_triplet\n",
    "df['transcript_name'] = df.gene_name+' '+df.transcript_triplet\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "34c115db-9cf8-4992-981e-930b0117c876",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ic_num</th>\n",
       "      <th>tss_num</th>\n",
       "      <th>tes_num</th>\n",
       "      <th>transcript_triplet</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>[3, 4, 5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>[4, 5, 4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>[1, 1, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000155801.2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>[1, 1, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>[2, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[2, 3, 2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  ic_num  tss_num  tes_num transcript_triplet\n",
       "0  ENSMUST00000153068.1       4        3        5          [3, 4, 5]\n",
       "1  ENSMUST00000146979.1       5        4        4          [4, 5, 4]\n",
       "2  ENSMUST00000155801.1       1        1        3          [1, 1, 3]\n",
       "3  ENSMUST00000155801.2       1        1        3          [1, 1, 3]\n",
       "4  ENSMUST00000179865.6       2        2        1          [2, 2, 1]\n",
       "5  ENSMUST00000020637.8       3        2        2          [2, 3, 2]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ic\n",
    "beds['tss']\n",
    "beds['tes']\n",
    "ttrip_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "763d217d-4d46-47cf-bbe5-2cc1a2763c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ad956e0c-12ac-4e6b-ba07-03c8782b6d21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hf = h5py.File('Canx_triplet.h5', 'w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1cb5d952-11b2-437b-a478-1700c9e986ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py:2446: PerformanceWarning: \n",
      "your performance may suffer as PyTables will pickle object types that it cannot\n",
      "map directly to c-types [inferred_type->mixed,key->block0_values] [items->Index(['chrom', 'strand', 'coordinates', 'transcript_id', 'ic_id'], dtype='object')]\n",
      "\n",
      "  encoding=encoding,\n"
     ]
    }
   ],
   "source": [
    "oname = 'Canx_triplet.h5'\n",
    "ic.to_hdf(oname, 'ic', mode='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d2238a52-92b9-428e-a71d-fdd0807cdf26",
   "metadata": {},
   "outputs": [],
   "source": [
    "beds['tss'].df.to_hdf(oname, 'tss', mode='a', format='table')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "003fa2f8-4ef1-449e-8699-06862c2c0fd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "beds['tes'].df.to_hdf(oname, 'tes', mode='a', format='table')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f5772e31-4b7e-4e07-859a-99f1a47f6abe",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttrip_df.to_hdf(oname, 'map', mode='a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "527eba46-e3e7-410e-85dd-9b3b7155004d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# peeps = h5py.File('Canx_triplet.h5', 'r')\n",
    "ic_2 = pd.read_hdf('Canx_triplet.h5', key='ic')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "40bfbf8b-5ce7-46e8-a44d-1812e5c55d38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>(50297363, 50298319)</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>ENSMUSG00000020368.14_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>(50298116, 50297273, 50297057, 50296198, 50299...</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>ENSMUSG00000020368.14_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>(50298837, 50300990)</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>ENSMUSG00000020368.14_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>(50311603, 50310793, 50310371, 50308787, 50308...</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>ENSMUSG00000020368.14_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>(50311603, 50310793, 50310371, 50308787, 50308...</td>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>ENSMUSG00000020368.14_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom strand                                        coordinates  \\\n",
       "0  chr11      -                               (50297363, 50298319)   \n",
       "1  chr11      -  (50298116, 50297273, 50297057, 50296198, 50299...   \n",
       "2  chr11      -                               (50298837, 50300990)   \n",
       "3  chr11      -  (50311603, 50310793, 50310371, 50308787, 50308...   \n",
       "4  chr11      -  (50311603, 50310793, 50310371, 50308787, 50308...   \n",
       "\n",
       "                               transcript_id                    ic_id  \n",
       "0                       ENSMUST00000153068.1  ENSMUSG00000020368.14_4  \n",
       "1                       ENSMUST00000146979.1  ENSMUSG00000020368.14_5  \n",
       "2  ENSMUST00000155801.1,ENSMUST00000155801.2  ENSMUSG00000020368.14_1  \n",
       "3                       ENSMUST00000179865.6  ENSMUSG00000020368.14_2  \n",
       "4                       ENSMUST00000020637.8  ENSMUSG00000020368.14_3  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ic_2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "72ef12e3-fcb9-41fd-a1ff-74dde9ff8ce0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50298268</td>\n",
       "      <td>50298369</td>\n",
       "      <td>ENSMUSG00000020368.14_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50299327</td>\n",
       "      <td>50299428</td>\n",
       "      <td>ENSMUSG00000020368.14_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50300939</td>\n",
       "      <td>50301040</td>\n",
       "      <td>ENSMUSG00000020368.14_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50325622</td>\n",
       "      <td>50325723</td>\n",
       "      <td>ENSMUSG00000020368.14_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome     Start       End                     Name Score Strand\n",
       "0      chr11  50298268  50298369  ENSMUSG00000020368.14_3     .      -\n",
       "1      chr11  50299327  50299428  ENSMUSG00000020368.14_4     .      -\n",
       "2      chr11  50300939  50301040  ENSMUSG00000020368.14_1     .      -\n",
       "3      chr11  50325622  50325723  ENSMUSG00000020368.14_2     .      -"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_bed_2 = pd.read_hdf('Canx_triplet.h5', key='tss')\n",
    "tss_bed_2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8253ec4f-7df2-490b-af5a-610c3b8eb514",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cerberus import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1a052999-7ab8-4e0c-bcea-a94caa3d5b92",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'tests/files/Canx.gtf'\n",
    "m = 'Canx_triplet_tid_map.tsv'\n",
    "agg = True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "69f4032b-d277-4865-9938-de7f457caba9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Chromosome   Source      Feature     Start       End Score Strand Frame  \\\n",
      "0       chr11  ENSEMBL   transcript  50293960  50325673     .      -     .   \n",
      "1       chr11  ENSEMBL         exon  50325492  50325673     .      -     .   \n",
      "2       chr11  ENSEMBL         exon  50311603  50311780     .      -     .   \n",
      "3       chr11  ENSEMBL          CDS  50311603  50311777     .      -     0   \n",
      "4       chr11  ENSEMBL  start_codon  50311774  50311777     .      -     0   \n",
      "..        ...      ...          ...       ...       ...   ...    ...   ...   \n",
      "82      chr11   HAVANA         exon  50298837  50299344     .      -     .   \n",
      "83      chr11   HAVANA          CDS  50300897  50300990     .      -     .   \n",
      "84      chr11   HAVANA   transcript  50298836  50300990     .      -     .   \n",
      "85      chr11   HAVANA         exon  50300897  50300990     .      -     .   \n",
      "86      chr11   HAVANA         exon  50298837  50299344     .      -     .   \n",
      "\n",
      "                  gene_id       gene_type  ...                 tag  \\\n",
      "0   ENSMUSG00000020368.14  protein_coding  ...                CCDS   \n",
      "1   ENSMUSG00000020368.14  protein_coding  ...                CCDS   \n",
      "2   ENSMUSG00000020368.14  protein_coding  ...                CCDS   \n",
      "3   ENSMUSG00000020368.14  protein_coding  ...                CCDS   \n",
      "4   ENSMUSG00000020368.14  protein_coding  ...                CCDS   \n",
      "..                    ...             ...  ...                 ...   \n",
      "82  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
      "83  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
      "84  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
      "85  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
      "86  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
      "\n",
      "   transcript_support_level       ccdsid exon_number               exon_id  \\\n",
      "0                         1  CCDS24633.1         NaN                   NaN   \n",
      "1                         1  CCDS24633.1           1  ENSMUSE00000371309.6   \n",
      "2                         1  CCDS24633.1           2  ENSMUSE00001016144.1   \n",
      "3                         1  CCDS24633.1           2  ENSMUSE00001016144.1   \n",
      "4                         1  CCDS24633.1           2  ENSMUSE00001016144.1   \n",
      "..                      ...          ...         ...                   ...   \n",
      "82                        3          NaN           2  ENSMUSE00000790569.1   \n",
      "83                        3          NaN           1  ENSMUSE00000776476.1   \n",
      "84                        3          NaN         NaN                   NaN   \n",
      "85                        3          NaN           1  ENSMUSE00000776476.1   \n",
      "86                        3          NaN           2  ENSMUSE00000790569.1   \n",
      "\n",
      "       havana_transcript                    transcript_id transcript_name  \\\n",
      "0                    NaN  ENSMUSG00000020368.14 [2, 2, 1]  Canx [2, 2, 1]   \n",
      "1                    NaN  ENSMUSG00000020368.14 [2, 2, 1]  Canx [2, 2, 1]   \n",
      "2                    NaN  ENSMUSG00000020368.14 [2, 2, 1]  Canx [2, 2, 1]   \n",
      "3                    NaN  ENSMUSG00000020368.14 [2, 2, 1]  Canx [2, 2, 1]   \n",
      "4                    NaN  ENSMUSG00000020368.14 [2, 2, 1]  Canx [2, 2, 1]   \n",
      "..                   ...                              ...             ...   \n",
      "82  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
      "83  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
      "84  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
      "85  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
      "86  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
      "\n",
      "                       original_transcript_id original_transcript_name  \n",
      "0                        ENSMUST00000179865.6                 Canx-201  \n",
      "1                        ENSMUST00000179865.6                 Canx-201  \n",
      "2                        ENSMUST00000179865.6                 Canx-201  \n",
      "3                        ENSMUST00000179865.6                 Canx-201  \n",
      "4                        ENSMUST00000179865.6                 Canx-201  \n",
      "..                                        ...                      ...  \n",
      "82  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
      "83  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
      "84  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
      "85  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
      "86  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
      "\n",
      "[87 rows x 26 columns]\n",
      "87\n",
      "87\n"
     ]
    }
   ],
   "source": [
    "df = pr.read_gtf(gtf).df\n",
    "m_df = pd.read_csv(m, sep='\\t')\n",
    "\n",
    "# groupby transcripts that are the same\n",
    "gb_cols = ['gene_name', 'gene_id', 'transcript_triplet',\n",
    "           'transcript_id', 'transcript_name']\n",
    "temp = m_df[['transcript_id',\n",
    "             'original_transcript_id',\n",
    "             'original_transcript_name']].copy(deep=True)\n",
    "m_df = m_df.groupby(gb_cols).agg({'original_transcript_id': ','.join,\n",
    "                                  'original_transcript_name': ','.join}).reset_index()\n",
    "m_df = m_df.merge(temp, on='transcript_id', suffixes=('','_merge'))\n",
    "m_df.drop(['gene_name', 'gene_id', 'transcript_triplet'],\n",
    "          axis=1, inplace=True)\n",
    "\n",
    "# add new transcript ids\n",
    "df = df.merge(m_df, left_on=['transcript_id', 'transcript_name'],\n",
    "              right_on=['original_transcript_id_merge',\n",
    "                        'original_transcript_name_merge'],\n",
    "             suffixes=('_x', ''))\n",
    "\n",
    "# drop old tids\n",
    "df.drop(['transcript_id_x', 'transcript_name_x',\n",
    "         'original_transcript_id_merge',\n",
    "         'original_transcript_name_merge'],\n",
    "        axis=1, inplace=True)\n",
    "\n",
    "# if we're only reporting one transcript per triplet\n",
    "print(df)\n",
    "if agg:\n",
    "    print(len(df.index))\n",
    "    df = df.loc[~(df.duplicated(subset=['transcript_id',\n",
    "                                         'transcript_name',\n",
    "                                         'Start', 'End',\n",
    "                                         'Feature'], keep='first'))|\\\n",
    "                ~((df.duplicated(subset=['transcript_id',\n",
    "                                         'transcript_name', \n",
    "                                         'Feature'], keep='first'))&\\\n",
    "                  (df.Feature=='transcript'))]\n",
    "    print(len(df.index))\n",
    "df = pr.PyRanges(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "279aa3f7-c3b0-4868-8085-d25f77feb155",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 2, 1]</td>\n",
       "      <td>Canx [2, 2, 1]</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>Canx-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50325492</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000371309.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 2, 1]</td>\n",
       "      <td>Canx [2, 2, 1]</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>Canx-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311780</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 2, 1]</td>\n",
       "      <td>Canx [2, 2, 1]</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>Canx-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 2, 1]</td>\n",
       "      <td>Canx [2, 2, 1]</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>Canx-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>50311774</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 2, 1]</td>\n",
       "      <td>Canx [2, 2, 1]</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>Canx-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298836</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>87 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | +21   |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+-------|\n",
       "| chr11        | ENSEMBL    | transcript   | 50293960  | 50325673  | ...   |\n",
       "| chr11        | ENSEMBL    | exon         | 50325492  | 50325673  | ...   |\n",
       "| chr11        | ENSEMBL    | exon         | 50311603  | 50311780  | ...   |\n",
       "| chr11        | ENSEMBL    | CDS          | 50311603  | 50311777  | ...   |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...   |\n",
       "| chr11        | HAVANA     | CDS          | 50300897  | 50300990  | ...   |\n",
       "| chr11        | HAVANA     | transcript   | 50298836  | 50300990  | ...   |\n",
       "| chr11        | HAVANA     | exon         | 50300897  | 50300990  | ...   |\n",
       "| chr11        | HAVANA     | exon         | 50298837  | 50299344  | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 87 rows and 26 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "21 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_status, level, ... (+ 14 more.)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c31dde6a-da72-4969-b53b-6fb278a5b605",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b9b6552d-0016-4e14-b609-a01486dea88d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298836</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50300897</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000776476.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50299344</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00000790569.1</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 1, 3]</td>\n",
       "      <td>Canx [1, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "79      chr11  HAVANA  transcript  50298837  50300990     .      -     .   \n",
       "80      chr11  HAVANA        exon  50300897  50300990     .      -     .   \n",
       "81      chr11  HAVANA         CDS  50300897  50300990     .      -     .   \n",
       "82      chr11  HAVANA        exon  50298837  50299344     .      -     .   \n",
       "83      chr11  HAVANA         CDS  50300897  50300990     .      -     .   \n",
       "84      chr11  HAVANA  transcript  50298836  50300990     .      -     .   \n",
       "85      chr11  HAVANA        exon  50300897  50300990     .      -     .   \n",
       "86      chr11  HAVANA        exon  50298837  50299344     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ...                 tag  \\\n",
       "79  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "80  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "81  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "82  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "83  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "84  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
       "85  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
       "86  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_1   \n",
       "\n",
       "   transcript_support_level ccdsid exon_number               exon_id  \\\n",
       "79                        3    NaN         NaN                   NaN   \n",
       "80                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "81                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "82                        3    NaN           2  ENSMUSE00000790569.1   \n",
       "83                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "84                        3    NaN         NaN                   NaN   \n",
       "85                        3    NaN           1  ENSMUSE00000776476.1   \n",
       "86                        3    NaN           2  ENSMUSE00000790569.1   \n",
       "\n",
       "       havana_transcript                    transcript_id transcript_name  \\\n",
       "79  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
       "80  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
       "81  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
       "82  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
       "83  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
       "84  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
       "85  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
       "86  OTTMUST00000012413.1  ENSMUSG00000020368.14 [1, 1, 3]  Canx [1, 1, 3]   \n",
       "\n",
       "                       original_transcript_id original_transcript_name  \n",
       "79  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "80  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "81  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "82  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "83  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "84  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "85  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "86  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "\n",
       "[8 rows x 26 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id == 'ENSMUSG00000020368.14 [1, 1, 3]']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b77a3c6a-d0fd-4bc7-960d-b90ddd22d5b2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
