{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "521791a1-c358-493d-97c3-22ffeb41d263",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ad850dca-ff08-4f1a-89e5-9ebf8734c4bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'tests/files/Canx.gtf'\n",
    "dist = 50\n",
    "\n",
    "gr_gtf = pr.read_gtf(gtf)\n",
    "\n",
    "df_tss = gr_gtf.features.tss()#.extend(dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "de6b1491-9939-4026-93b2-8ab0285ba51f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>tss</td>\n",
       "      <td>50325672</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>tss</td>\n",
       "      <td>50325672</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-001</td>\n",
       "      <td>ENSMUSP00000020637.8</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012410.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>tss</td>\n",
       "      <td>50299377</td>\n",
       "      <td>50299378</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012412.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>tss</td>\n",
       "      <td>50298318</td>\n",
       "      <td>50298319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>tss</td>\n",
       "      <td>50300989</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Source     | Feature    |     Start |       End | Score      | Strand       | +17   |\n",
       "| (category)   | (object)   | (object)   |   (int32) |   (int32) | (object)   | (category)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr11        | ENSEMBL    | tss        |  50325672 |  50325673 | .          | -            | ...   |\n",
       "| chr11        | HAVANA     | tss        |  50325672 |  50325673 | .          | -            | ...   |\n",
       "| chr11        | HAVANA     | tss        |  50299377 |  50299378 | .          | -            | ...   |\n",
       "| chr11        | HAVANA     | tss        |  50298318 |  50298319 | .          | -            | ...   |\n",
       "| chr11        | HAVANA     | tss        |  50300989 |  50300990 | .          | -            | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 5 rows and 24 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Frame, gene_id, gene_type, gene_status, level, havana_gene, transcript_id, transcript_type, ... (+ 9 more.)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tss.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0319f79f-4b3a-4ee1-a2f5-881347e62b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = {'Chromosome': ['chr1', 'chr1', 'chr1'], 'Start': [3, 8, 5], 'End': [6, 9, 7],\n",
    "     'Strand': ['+', '+', '-']}\n",
    "gr = pr.from_dict(d)\n",
    "gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f6e2c6da-c9a9-4043-a998-2feb0f8cdf70",
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_bed = 'tests/files/Canx_tss.bed'\n",
    "tes_bed = 'tests/files/Canx_tes.bed'\n",
    "gtf = 'tests/files/Canx.gtf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "657c0947-23a9-4c29-a114-1404b3a174a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = pr.read_gtf(gtf, duplicate_attr=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "93120be7-6cf8-4f96-9b63-73fddd0ba5f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50325492</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000371309.6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311780</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>50311774</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | Score      | Strand       | +17   |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) | (object)   | (category)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+-------|\n",
       "| chr11        | HAVANA     | gene         |  50293960 |  50325673 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | transcript   |  50293960 |  50325673 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50325492 |  50325673 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50311603 |  50311780 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | CDS          |  50311603 |  50311777 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | start_codon  |  50311774 |  50311777 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50310793 |  50310867 | .          | -            | ...   |\n",
       "| chr11        | ENSEMBL    | CDS          |  50310793 |  50310867 | .          | -            | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+-------+\n",
       "Stranded PyRanges object has 8 rows and 24 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "17 hidden columns: Frame, gene_id, gene_type, gene_status, level, havana_gene, transcript_id, transcript_type, ... (+ 9 more.)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "90e2a7c7-2006-4a83-87a4-0335052cf8b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = gtf.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "be7653a4-03cd-4f71-8ad7-bde3d009f797",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = gtf.loc[gtf.Feature == 'transcript']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a6ed3fda-e718-4ce3-8a7f-2bed56cd475f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50294466</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-001</td>\n",
       "      <td>ENSMUSP00000020637.8</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012410.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50296198</td>\n",
       "      <td>50299378</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012412.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50297363</td>\n",
       "      <td>50298319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome   Source     Feature     Start       End Score Strand Frame  \\\n",
       "1       chr11  ENSEMBL  transcript  50293960  50325673     .      -     .   \n",
       "36      chr11   HAVANA  transcript  50294466  50325673     .      -     .   \n",
       "71      chr11   HAVANA  transcript  50296198  50299378     .      -     .   \n",
       "77      chr11   HAVANA  transcript  50297363  50298319     .      -     .   \n",
       "80      chr11   HAVANA  transcript  50298837  50300990     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ...       transcript_type  \\\n",
       "1   ENSMUSG00000020368.14  protein_coding  ...        protein_coding   \n",
       "36  ENSMUSG00000020368.14  protein_coding  ...        protein_coding   \n",
       "71  ENSMUSG00000020368.14  protein_coding  ...  processed_transcript   \n",
       "77  ENSMUSG00000020368.14  protein_coding  ...  processed_transcript   \n",
       "80  ENSMUSG00000020368.14  protein_coding  ...  processed_transcript   \n",
       "\n",
       "   transcript_status transcript_name            protein_id  \\\n",
       "1              KNOWN        Canx-201  ENSMUSP00000137440.1   \n",
       "36             KNOWN        Canx-001  ENSMUSP00000020637.8   \n",
       "71             KNOWN        Canx-003                   NaN   \n",
       "77             KNOWN        Canx-002                   NaN   \n",
       "80             KNOWN        Canx-004                   NaN   \n",
       "\n",
       "                              tag transcript_support_level       ccdsid  \\\n",
       "1   basic,appris_principal_1,CCDS                        1  CCDS24633.1   \n",
       "36  basic,appris_principal_1,CCDS                        1  CCDS24633.1   \n",
       "71                            NaN                        2          NaN   \n",
       "77                            NaN                        5          NaN   \n",
       "80                            NaN                        3          NaN   \n",
       "\n",
       "   exon_number exon_id     havana_transcript  \n",
       "1          NaN     NaN                   NaN  \n",
       "36         NaN     NaN  OTTMUST00000012410.1  \n",
       "71         NaN     NaN  OTTMUST00000012412.1  \n",
       "77         NaN     NaN  OTTMUST00000012411.1  \n",
       "80         NaN     NaN  OTTMUST00000012413.1  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "45b628dc-6f5a-4b0f-8b7d-e053e0f15b53",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf['basic_set'] = gtf.tag.str.contains('basic')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "12b1093c-f446-4408-ab45-c901de7bf53f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf['temp'] = gtf.tag.str.split('appris_principal_', n=1, expand=True)[1]\n",
    "gtf['appris_principal'] = gtf.temp.str.split(',', n=1, expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "972aa359-e0b4-4725-8389-b1cadbe4fdcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf.basic_set.fillna(False, inplace=True)\n",
    "gtf = gtf.sort_values(by=['appris_principal', 'basic_set'], \n",
    "                      ascending=[False, True],\n",
    "                      na_position='last')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "0abfe0c1-3e20-4c79-97eb-f84bd97dbf5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>basic_set</th>\n",
       "      <th>temp</th>\n",
       "      <th>appris_principal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>1,CCDS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50294466</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSP00000020637.8</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012410.1</td>\n",
       "      <td>True</td>\n",
       "      <td>1,CCDS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50296198</td>\n",
       "      <td>50299378</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012412.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50297363</td>\n",
       "      <td>50298319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome   Source     Feature     Start       End Score Strand Frame  \\\n",
       "1       chr11  ENSEMBL  transcript  50293960  50325673     .      -     .   \n",
       "36      chr11   HAVANA  transcript  50294466  50325673     .      -     .   \n",
       "71      chr11   HAVANA  transcript  50296198  50299378     .      -     .   \n",
       "77      chr11   HAVANA  transcript  50297363  50298319     .      -     .   \n",
       "80      chr11   HAVANA  transcript  50298837  50300990     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ...            protein_id  \\\n",
       "1   ENSMUSG00000020368.14  protein_coding  ...  ENSMUSP00000137440.1   \n",
       "36  ENSMUSG00000020368.14  protein_coding  ...  ENSMUSP00000020637.8   \n",
       "71  ENSMUSG00000020368.14  protein_coding  ...                   NaN   \n",
       "77  ENSMUSG00000020368.14  protein_coding  ...                   NaN   \n",
       "80  ENSMUSG00000020368.14  protein_coding  ...                   NaN   \n",
       "\n",
       "                              tag transcript_support_level       ccdsid  \\\n",
       "1   basic,appris_principal_1,CCDS                        1  CCDS24633.1   \n",
       "36  basic,appris_principal_1,CCDS                        1  CCDS24633.1   \n",
       "71                            NaN                        2          NaN   \n",
       "77                            NaN                        5          NaN   \n",
       "80                            NaN                        3          NaN   \n",
       "\n",
       "   exon_number exon_id     havana_transcript basic_set    temp  \\\n",
       "1          NaN     NaN                   NaN      True  1,CCDS   \n",
       "36         NaN     NaN  OTTMUST00000012410.1      True  1,CCDS   \n",
       "71         NaN     NaN  OTTMUST00000012412.1     False     NaN   \n",
       "77         NaN     NaN  OTTMUST00000012411.1     False     NaN   \n",
       "80         NaN     NaN  OTTMUST00000012413.1     False     NaN   \n",
       "\n",
       "   appris_principal  \n",
       "1                 1  \n",
       "36                1  \n",
       "71              NaN  \n",
       "77              NaN  \n",
       "80              NaN  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb0f1756-e68e-434f-8560-086a7945b188",
   "metadata": {},
   "source": [
    "## compute triplets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "003d5746-81e3-4c99-b32c-ef798294b716",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transcript_ref(fname):\n",
    "    \n",
    "    df = pr.read_gtf(fname, duplicate_attr=True).df\n",
    "    df = df.loc[df.Feature == 'transcript']\n",
    "    df['basic_set'] = df.tag.str.contains('basic')\n",
    "    df['temp'] = df.tag.str.split('appris_principal_', n=1, expand=True)[1]\n",
    "    df['appris_principal'] = df.temp.str.split(',', n=1, expand=True)[0]\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "id": "dbf0861b-8d52-4452-bafa-75a5434f61bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def number_tss_ic_tes(df, mode):\n",
    "    # groupby feature but record which feature\n",
    "    # each transcript id uses \n",
    "    cols = ['transcript_id', mode, 'basic_set', 'appris_principal', 'gene_id']\n",
    "    df = df[cols].groupby([mode, 'gene_id']).agg({'transcript_id': ','.join,\n",
    "                                     'basic_set': 'max',\n",
    "                                     'appris_principal': 'min'}).reset_index()\n",
    "\n",
    "    # compute intron chain number\n",
    "    df['{}_num'.format(mode)] = df.sort_values(by=['gene_id', 'appris_principal', 'basic_set'],\n",
    "                                 ascending=[True, False, True],\n",
    "                                 na_position='last')\\\n",
    "                                 .groupby(['gene_id'])\\\n",
    "                                 .cumcount() + 1\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "id": "b8c2ab21-29ee-498d-aeef-604be3aebd92",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = 'tests/files/Canx.gtf'\n",
    "tss_bed = 'tests/files/Canx_tss.bed'\n",
    "tes_bed = 'tests/files/Canx_tes.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "id": "33020490-415a-4ae3-ab13-43ac345164ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "### intron chain annotation ###\n",
    "\n",
    "# get basic status and appris_principal tag for each transcript\n",
    "t_df = get_transcript_ref(fname)\n",
    "t_df = t_df[['transcript_id', 'gene_id', 'basic_set', 'appris_principal']]\n",
    "\n",
    "# get unique intron chains from gtf\n",
    "df = pr.read_gtf(fname).df\n",
    "\n",
    "# restrict to exon entries\n",
    "df = df.loc[df.Feature == 'exon']\n",
    "cols = ['Chromosome', 'Strand', 'Start', 'End', 'transcript_id', 'gene_id']\n",
    "df = df[cols]\n",
    "\n",
    "# melt to isolate individual coordinates\n",
    "df = pd.melt(df, id_vars=['Chromosome', 'Strand', 'transcript_id', 'gene_id'],\n",
    "            value_vars=['Start', 'End'],\n",
    "            value_name='Coord')\n",
    "df.drop('variable', axis=1, inplace=True)\n",
    "\n",
    "# sort to order coordinates correctly\n",
    "df.Coord = df.Coord.astype(int)\n",
    "fwd = df.loc[df.Strand == '+'].copy(deep=True)\n",
    "rev = df.loc[df.Strand == '-'].copy(deep=True)\n",
    "\n",
    "fwd.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                ascending=[True, True, True])\n",
    "rev.sort_values(by=['Chromosome', 'transcript_id', 'Coord'],\n",
    "                ascending=[True, True, False])\n",
    "df = pd.concat([fwd, rev])\n",
    "\n",
    "# create intron chain strings\n",
    "df.Coord = df.Coord.astype(str)\n",
    "df = df.groupby(['Chromosome', 'Strand', 'transcript_id'])['Coord'].apply('-'.join).reset_index()\n",
    "\n",
    "# remove tss and tes from intron chain\n",
    "df['temp'] = df.Coord.str.split('-', n=1, expand=True)[1]\n",
    "df['ic'] = df.temp.str.rsplit('-', n=1, expand=True)[0]\n",
    "\n",
    "# add basic annotation, appris principal number, and gene id\n",
    "df = df.merge(t_df, on='transcript_id', how='left')\n",
    "\n",
    "# add number for each unique intron chain\n",
    "df = number_tss_ic_tes(df, mode='ic')\n",
    "\n",
    "# keep track of which transcripts use each intron chain\n",
    "df.transcript_id = df.transcript_id.str.split(',')\n",
    "ttrip_df = df[['transcript_id', 'ic_num']].explode(column='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "id": "b82ba4e2-8ed5-41bb-b99f-b9028177d7c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### tss / tes numbering ###\n",
    "beds = {}\n",
    "for mode, bed_file in zip(['tss', 'tes'], [tss_bed, tes_bed]):\n",
    "\n",
    "    ends = pr.read_bed(bed_file).df\n",
    "    ends.rename({'ThickStart': 'gene_id'}, axis=1, inplace=True)\n",
    "    ends[mode] = [i for i in range(len(ends.index))]\n",
    "    ends = pr.PyRanges(ends)\n",
    "\n",
    "    # merge gtf ends with bed ends\n",
    "    df = pr.read_gtf(gtf)\n",
    "    if mode == 'tss':\n",
    "        df = df.features.tss().df\n",
    "    elif mode == 'tes':\n",
    "        df = df.features.tes().df\n",
    "\n",
    "    # merge in information about transcript status\n",
    "    t_df = get_transcript_ref(gtf)\n",
    "    t_df = t_df[['transcript_id', 'appris_principal', 'basic_set']]\n",
    "    df = df.merge(t_df, how='left', on='transcript_id')\n",
    "    df = pr.PyRanges(df)\n",
    "\n",
    "    # join with bed ends\n",
    "    df = df.join(ends, \n",
    "                 strandedness='same',\n",
    "                 how='left').df\n",
    "    df = df.loc[df.gene_id == df.gene_id_b]\n",
    "\n",
    "    # add number for each unique end region\n",
    "    df = number_tss_ic_tes(df, mode=mode)\n",
    "    \n",
    "    # keep track of which transcripts use each intron chain\n",
    "    df.transcript_id = df.transcript_id.str.split(',')\n",
    "    temp = df[['transcript_id', '{}_num'.format(mode)]].explode(column='transcript_id')  \n",
    "    ttrip_df = ttrip_df.merge(temp, on='transcript_id')   \n",
    "    \n",
    "    # add numbers back into bed file\n",
    "    ends = ends.df\n",
    "    ends = ends.merge(df[[mode, '{}_num'.format(mode)]], on=mode)\n",
    "    \n",
    "    ends['Name'] = ends.gene_id+'_'+ends['{}_num'.format(mode)].astype(str)\n",
    "    ends.drop([mode, 'gene_id', '{}_num'.format(mode)], axis=1, inplace=True)\n",
    "    beds[mode] = pr.PyRanges(ends)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "id": "d4e810df-c35b-43e9-858d-32dd3951292c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create formatted string to represent each transcript\n",
    "ttrip_df['transcript_triplet'] = '['+ttrip_df['tss_num'].astype(str)+', '+\\\n",
    "                                     ttrip_df['ic_num'].astype(str)+', '+\\\n",
    "                                     ttrip_df['tes_num'].astype(str)+']'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "59900b0b-a160-4684-b57f-ab69922009bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ic_num</th>\n",
       "      <th>tss_num</th>\n",
       "      <th>tes_num</th>\n",
       "      <th>transcript_triplet</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>[4, 4, 5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>[3, 5, 4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>[2, 1, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000155801.2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>[2, 1, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>[1, 3, 2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  ic_num  tss_num  tes_num transcript_triplet\n",
       "0  ENSMUST00000153068.1       4        4        5          [4, 4, 5]\n",
       "1  ENSMUST00000146979.1       5        3        4          [3, 5, 4]\n",
       "2  ENSMUST00000155801.1       1        2        3          [2, 1, 3]\n",
       "3  ENSMUST00000155801.2       1        2        3          [2, 1, 3]\n",
       "4  ENSMUST00000179865.6       2        1        1          [1, 2, 1]\n",
       "5  ENSMUST00000020637.8       3        1        2          [1, 3, 2]"
      ]
     },
     "execution_count": 354,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ttrip_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "id": "00935205-a74a-4f3c-aa64-6eaa3755a825",
   "metadata": {},
   "outputs": [],
   "source": [
    "### output files ###\n",
    "\n",
    "# gtf\n",
    "df = pr.read_gtf(gtf, duplicate_attr=True).df\n",
    "df = df.merge(ttrip_df[['transcript_id', 'transcript_triplet']], \n",
    "              how='left',\n",
    "              on='transcript_id')\n",
    "df = pr.PyRanges(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "id": "a371d9ba-f01b-485b-ae1a-e80cf99c62dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50325622</td>\n",
       "      <td>50325723</td>\n",
       "      <td>ENSMUSG00000020368.14_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50299327</td>\n",
       "      <td>50299428</td>\n",
       "      <td>ENSMUSG00000020368.14_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50298268</td>\n",
       "      <td>50298369</td>\n",
       "      <td>ENSMUSG00000020368.14_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>50300939</td>\n",
       "      <td>50301040</td>\n",
       "      <td>ENSMUSG00000020368.14_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------------------+------------+--------------+\n",
       "| Chromosome   |     Start |       End | Name                    | Score      | Strand       |\n",
       "| (category)   |   (int32) |   (int32) | (object)                | (object)   | (category)   |\n",
       "|--------------+-----------+-----------+-------------------------+------------+--------------|\n",
       "| chr11        |  50325622 |  50325723 | ENSMUSG00000020368.14_1 | .          | -            |\n",
       "| chr11        |  50299327 |  50299428 | ENSMUSG00000020368.14_3 | .          | -            |\n",
       "| chr11        |  50298268 |  50298369 | ENSMUSG00000020368.14_4 | .          | -            |\n",
       "| chr11        |  50300939 |  50301040 | ENSMUSG00000020368.14_2 | .          | -            |\n",
       "+--------------+-----------+-----------+-------------------------+------------+--------------+\n",
       "Stranded PyRanges object has 4 rows and 6 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 356,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beds['tss']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "id": "2ddeb85b-5531-4f3a-a720-f779749b640d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>transcript_triplet</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50325492</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000371309.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311780</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50311603</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>50311774</td>\n",
       "      <td>50311777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001016144.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50310793</td>\n",
       "      <td>50310867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>ENSMUSP00000137440.1</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSE00000103860.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 2, 1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+-------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | Score      | +19   |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) | (object)   | ...   |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+-------|\n",
       "| chr11        | HAVANA     | gene         |  50293960 |  50325673 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | transcript   |  50293960 |  50325673 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50325492 |  50325673 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50311603 |  50311780 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | CDS          |  50311603 |  50311777 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | start_codon  |  50311774 |  50311777 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | exon         |  50310793 |  50310867 | .          | ...   |\n",
       "| chr11        | ENSEMBL    | CDS          |  50310793 |  50310867 | .          | ...   |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+-------+\n",
       "Stranded PyRanges object has 8 rows and 25 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "19 hidden columns: Strand, Frame, gene_id, gene_type, gene_status, level, havana_gene, transcript_id, ... (+ 11 more.)"
      ]
     },
     "execution_count": 338,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "id": "f3309e1d-2e1e-44b1-90d1-c391a4a06ac2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "94c00c48-876a-4803-bd47-65f5846a8ba7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ic</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>basic_set</th>\n",
       "      <th>appris_principal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>50297363-50298319</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>50298116-50297273-50297057-50296198-50299378-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>50298837-50300990</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>50311603-50310793-50310371-50308787-50308309-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>50311603-50310793-50310371-50308787-50308309-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  ic                gene_id  \\\n",
       "0                                  50297363-50298319  ENSMUSG00000020368.14   \n",
       "1  50298116-50297273-50297057-50296198-50299378-5...  ENSMUSG00000020368.14   \n",
       "2                                  50298837-50300990  ENSMUSG00000020368.14   \n",
       "3  50311603-50310793-50310371-50308787-50308309-5...  ENSMUSG00000020368.14   \n",
       "4  50311603-50310793-50310371-50308787-50308309-5...  ENSMUSG00000020368.14   \n",
       "\n",
       "                               transcript_id basic_set appris_principal  \n",
       "0                       ENSMUST00000153068.1       NaN              NaN  \n",
       "1                       ENSMUST00000146979.1       NaN              NaN  \n",
       "2  ENSMUST00000155801.1,ENSMUST00000155801.2      True                1  \n",
       "3                       ENSMUST00000179865.6      True                1  \n",
       "4                       ENSMUST00000020637.8      True                1  "
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "c64e4149-01f0-46fc-992f-a774b84a7945",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "eb44df97-2164-4160-9960-39670f6bd887",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ic</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>basic_set</th>\n",
       "      <th>appris_principal</th>\n",
       "      <th>ic_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>50297363-50298319</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>50298116-50297273-50297057-50296198-50299378-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>50298837-50300990</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>50311603-50310793-50310371-50308787-50308309-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>50311603-50310793-50310371-50308787-50308309-5...</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  ic                gene_id  \\\n",
       "0                                  50297363-50298319  ENSMUSG00000020368.14   \n",
       "1  50298116-50297273-50297057-50296198-50299378-5...  ENSMUSG00000020368.14   \n",
       "2                                  50298837-50300990  ENSMUSG00000020368.14   \n",
       "3  50311603-50310793-50310371-50308787-50308309-5...  ENSMUSG00000020368.14   \n",
       "4  50311603-50310793-50310371-50308787-50308309-5...  ENSMUSG00000020368.14   \n",
       "\n",
       "                               transcript_id basic_set appris_principal  \\\n",
       "0                       ENSMUST00000153068.1       NaN              NaN   \n",
       "1                       ENSMUST00000146979.1       NaN              NaN   \n",
       "2  ENSMUST00000155801.1,ENSMUST00000155801.2      True                1   \n",
       "3                       ENSMUST00000179865.6      True                1   \n",
       "4                       ENSMUST00000020637.8      True                1   \n",
       "\n",
       "   ic_num  \n",
       "0       4  \n",
       "1       5  \n",
       "2       1  \n",
       "3       2  \n",
       "4       3  "
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "018e6889-01e6-4db1-bdca-ea337fe2aefa",
   "metadata": {},
   "source": [
    "#### make canx abundance file for testing purposes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "id": "6c1b35a5-a01c-4260-8c13-7b1fe6cfe339",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/Users/fairliereese/mortazavi_lab/data/mousewg/lr_bulk/talon/mouse_talon_abundance.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 418,
   "id": "72676460-b8df-43b6-b87a-c4a827d4bf1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.loc[df.annot_gene_name == 'Canx']\n",
    "df = df.loc[df.transcript_novelty == 'Known']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 419,
   "id": "10070474-4b6a-49e3-a9d4-d87b4b19d65c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename transcripts according to transcript names\n",
    "gtf_df = pr.read_gtf('tests/files/Canx.gtf').df\n",
    "gtf_df = gtf_df.loc[gtf_df.Feature == 'transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 420,
   "id": "26a64c27-e35c-4ef2-a67d-a191e6ac2e10",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df.transcript_id\n",
    "gtf_df['tid_stable'] = gtf_df.transcript_id.str.split('.', n=1, expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 421,
   "id": "d3ebd5f7-a27c-4d4a-9ee5-331280014393",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['tid_stable'] = df.annot_transcript_id.str.split('.', n=1, expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 422,
   "id": "6e4e8ac8-e431-40d0-b8a6-c5315eebe979",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = gtf_df[['tid_stable', 'transcript_id']]\n",
    "df = df.merge(gtf_df, on='tid_stable')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 423,
   "id": "cf8bcbb7-39db-4880-ab38-3b425efdb2f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.annot_transcript_id = df.transcript_id\n",
    "df.drop(['transcript_id', 'tid_stable'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 424,
   "id": "105abb94-3272-44b0-8f5b-fd2bfe3f6dbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>gastroc_25d_m_2</th>\n",
       "      <th>gastroc_25d_m_1</th>\n",
       "      <th>gastroc_10d_f_2</th>\n",
       "      <th>gastroc_10d_f_1</th>\n",
       "      <th>gastroc_25d_f_1</th>\n",
       "      <th>gastroc_25d_f_2</th>\n",
       "      <th>gastroc_4d_m_1</th>\n",
       "      <th>gastroc_4d_m_2</th>\n",
       "      <th>gastroc_10d_m_1</th>\n",
       "      <th>gastroc_10d_m_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>32839</td>\n",
       "      <td>86107</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-205</td>\n",
       "      <td>15</td>\n",
       "      <td>4281</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32839</td>\n",
       "      <td>86108</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-201</td>\n",
       "      <td>15</td>\n",
       "      <td>3779</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>44</td>\n",
       "      <td>26</td>\n",
       "      <td>67</td>\n",
       "      <td>38</td>\n",
       "      <td>35</td>\n",
       "      <td>104</td>\n",
       "      <td>96</td>\n",
       "      <td>60</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>32839</td>\n",
       "      <td>86109</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-202</td>\n",
       "      <td>5</td>\n",
       "      <td>708</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32839</td>\n",
       "      <td>86110</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-203</td>\n",
       "      <td>2</td>\n",
       "      <td>234</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32839</td>\n",
       "      <td>86111</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000155801.1</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-204</td>\n",
       "      <td>2</td>\n",
       "      <td>600</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>32839</td>\n",
       "      <td>86111</td>\n",
       "      <td>ENSMUSG00000020368.15</td>\n",
       "      <td>ENSMUST00000155801.2</td>\n",
       "      <td>Canx</td>\n",
       "      <td>Canx-204</td>\n",
       "      <td>2</td>\n",
       "      <td>600</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 102 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_ID  transcript_ID          annot_gene_id   annot_transcript_id  \\\n",
       "0    32839          86107  ENSMUSG00000020368.15  ENSMUST00000179865.6   \n",
       "1    32839          86108  ENSMUSG00000020368.15  ENSMUST00000020637.8   \n",
       "2    32839          86109  ENSMUSG00000020368.15  ENSMUST00000146979.1   \n",
       "3    32839          86110  ENSMUSG00000020368.15  ENSMUST00000153068.1   \n",
       "4    32839          86111  ENSMUSG00000020368.15  ENSMUST00000155801.1   \n",
       "5    32839          86111  ENSMUSG00000020368.15  ENSMUST00000155801.2   \n",
       "\n",
       "  annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "0            Canx              Canx-205       15    4281        Known   \n",
       "1            Canx              Canx-201       15    3779        Known   \n",
       "2            Canx              Canx-202        5     708        Known   \n",
       "3            Canx              Canx-203        2     234        Known   \n",
       "4            Canx              Canx-204        2     600        Known   \n",
       "5            Canx              Canx-204        2     600        Known   \n",
       "\n",
       "  transcript_novelty  ... gastroc_25d_m_2  gastroc_25d_m_1  gastroc_10d_f_2  \\\n",
       "0              Known  ...               1                0                0   \n",
       "1              Known  ...              12               44               26   \n",
       "2              Known  ...               1                1                1   \n",
       "3              Known  ...               0                0                0   \n",
       "4              Known  ...               0                0                0   \n",
       "5              Known  ...               0                0                0   \n",
       "\n",
       "   gastroc_10d_f_1  gastroc_25d_f_1  gastroc_25d_f_2  gastroc_4d_m_1  \\\n",
       "0                3                0                1               4   \n",
       "1               67               38               35             104   \n",
       "2                1                0                2               2   \n",
       "3                0                0                0               0   \n",
       "4                0                0                0               0   \n",
       "5                0                0                0               0   \n",
       "\n",
       "   gastroc_4d_m_2  gastroc_10d_m_1  gastroc_10d_m_2  \n",
       "0               2                0                0  \n",
       "1              96               60               34  \n",
       "2               1                3                0  \n",
       "3               0                0                0  \n",
       "4               0                0                0  \n",
       "5               0                0                0  \n",
       "\n",
       "[6 rows x 102 columns]"
      ]
     },
     "execution_count": 424,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "id": "4294d336-7070-4af9-88e3-006accfbe8b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('tests/files/Canx_abundance.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eee194e-e209-4da4-8cdc-44f2f69ba314",
   "metadata": {},
   "source": [
    "#### Changing transcript IDs in TALON abundance file and optionally? aggregating counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 591,
   "id": "4e6045ba-5775-4e5c-b95c-547c324fc149",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nov_ranks():\n",
    "    nov_rank = {'Known': 0,\n",
    "           'NIC': 1,\n",
    "           'ISM': 2,\n",
    "           'NNC': 3,\n",
    "           'Genomic': 4,\n",
    "           'Antisense': 5,\n",
    "           'Intergenic': 6}\n",
    "    \n",
    "    rank_nov = dict()\n",
    "    for key, item in nov_rank.items():\n",
    "        rank_nov[item] = key\n",
    "\n",
    "    return nov_rank, rank_nov\n",
    "\n",
    "def get_ism_ranks():\n",
    "    ism_rank = {'Both': 0,\n",
    "                'Prefix': 1,\n",
    "                'Suffix': 2,\n",
    "                None: 3}\n",
    "    rank_ism = dict()\n",
    "    for key, item in ism_rank.items():\n",
    "        rank_ism[item] = key\n",
    "\n",
    "    return ism_rank, rank_ism\n",
    "\n",
    "def get_non_dataset_cols():\n",
    "    non_dataset_cols = ['gene_ID', 'transcript_ID', 'annot_gene_id',\n",
    "                   'annot_transcript_id', 'annot_gene_name',\n",
    "                   'annot_transcript_name', 'n_exons', 'length',\n",
    "                   'gene_novelty', 'transcript_novelty', 'ISM_subtype']\n",
    "    return non_dataset_cols\n",
    "\n",
    "def get_dataset_cols(df):\n",
    "    \"\"\"\n",
    "    Get a list of the dataset columns\n",
    "\n",
    "    \"\"\"\n",
    "    non_dataset_cols = get_non_dataset_cols()\n",
    "    dataset_cols = [ x for x in list(df.columns) \\\n",
    "                        if x not in non_dataset_cols ]\n",
    "    return dataset_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 601,
   "id": "2d32db35-897f-4b5f-ae79-061a32aadd2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = 'tests/files/Canx_abundance.tsv'\n",
    "gtf = 'tests/files/Canx.gtf'\n",
    "m = 'tests/files/Canx_triplet_tid_map.tsv'\n",
    "h5 = '' # option to use / save in previous step as h5df?\n",
    "\n",
    "agg = True\n",
    "\n",
    "\n",
    "df = pd.read_csv(ab, sep='\\t')\n",
    "m_df = pd.read_csv(m, sep='\\t')\n",
    "\n",
    "# fix transcript ids in abundance file\n",
    "ab_map = m_df[['original_transcript_id', 'original_transcript_name', 'transcript_name', 'transcript_id']]\n",
    "df = df.merge(ab_map, left_on='annot_transcript_id', right_on='original_transcript_id')\n",
    "\n",
    "df.drop(['original_transcript_id', 'original_transcript_name', \n",
    "         'annot_transcript_id', 'annot_transcript_name'],\n",
    "        axis=1, inplace=True)\n",
    "df.rename({'transcript_name': 'annot_transcript_name',\n",
    "           'transcript_id': 'annot_transcript_id'},\n",
    "          axis=1, inplace=True)\n",
    "\n",
    "# aggregate counts if requested\n",
    "if agg:\n",
    "    gb_cols = ['gene_ID', 'annot_gene_id', 'annot_gene_name',\n",
    "               'gene_novelty', 'annot_transcript_name',\n",
    "               'annot_transcript_id']\n",
    "    \n",
    "    # handle properties which won't always correspond across the transcripts\n",
    "    # these are all subject to change\n",
    "    nov_rank, rank_nov = get_nov_ranks()\n",
    "    df['nov_rank'] = df.transcript_novelty.map(nov_rank)\n",
    "    df.drop('transcript_novelty', axis=1, inplace=True)\n",
    "    \n",
    "    ism_rank, rank_ism = get_ism_ranks()\n",
    "    df['ism_rank'] = df.ISM_subtype.map(ism_rank)\n",
    "    df.drop('ISM_subtype', axis=1, inplace=True)\n",
    "    \n",
    "    df['transcript_ID'] = df.transcript_ID.astype(str)\n",
    "    agg_dict = {'transcript_ID': ','.join,\n",
    "                'n_exons': 'mean', \n",
    "                'length': 'mean',\n",
    "                'nov_rank': 'min',\n",
    "                'ism_rank': 'min'}\n",
    "    cols = gb_cols + list(agg_dict.keys())\n",
    "    for c in list(set(df.columns)-set(cols)):\n",
    "        agg_dict[c] = 'sum'\n",
    "    \n",
    "    df = df.groupby(gb_cols).agg(agg_dict).reset_index()\n",
    "    df['transcript_novelty'] = df.nov_rank.map(rank_nov)\n",
    "    df.drop(['nov_rank'], axis=1, inplace=True)\n",
    "    df['ISM_subtype'] = df.ism_rank.map(rank_ism)\n",
    "    df.drop(['ism_rank'], axis=1, inplace=True)\n",
    "    \n",
    "# reorder columns\n",
    "c1 = get_non_dataset_cols()\n",
    "c2 = get_dataset_cols(df)\n",
    "df = df[c1+c2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 602,
   "id": "bbd6938f-2c95-4c18-b1fa-b6a7ca37a43b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.read_gtf(gtf).df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 603,
   "id": "929512d7-d20a-4595-9f87-8f0ff8fe4111",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_df.head()\n",
    "\n",
    "\n",
    "# groupby transcripts that are the same\n",
    "gb_cols = ['gene_name', 'gene_id', 'transcript_triplet',\n",
    "           'transcript_id', 'transcript_name']\n",
    "temp = m_df[['transcript_id',\n",
    "             'original_transcript_id',\n",
    "             'original_transcript_name']].copy(deep=True)\n",
    "m_df = m_df.groupby(gb_cols).agg({'original_transcript_id': ','.join,\n",
    "                                  'original_transcript_name': ','.join}).reset_index()\n",
    "m_df = m_df.merge(temp, on='transcript_id', suffixes=('','_merge'))\n",
    "m_df.drop(['gene_name', 'gene_id', 'transcript_triplet'],\n",
    "          axis=1, inplace=True)\n",
    "\n",
    "# add new transcript ids \n",
    "df = df.merge(m_df, left_on=['transcript_id', 'transcript_name'],\n",
    "              right_on=['original_transcript_id_merge',\n",
    "                        'original_transcript_name_merge'],\n",
    "             suffixes=('_x', ''))\n",
    "\n",
    "# drop old tids\n",
    "df.drop(['transcript_id_x', 'transcript_name_x',\n",
    "         'original_transcript_id_merge',\n",
    "         'original_transcript_name_merge'],\n",
    "        axis=1, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 621,
   "id": "bb45f7ee-4194-4866-aa3a-8ba3f8b6f92e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "85\n"
     ]
    }
   ],
   "source": [
    "print(len(df.index))\n",
    "if agg:\n",
    "    df = df.loc[~((df.duplicated(subset=['transcript_id',\n",
    "                                         'transcript_name',\n",
    "                                         'exon_id']))&(df.Feature.isin(['transcript', 'exon'])))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 622,
   "id": "f406f7ec-6bb7-446c-9653-14983d833652",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50293960</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 2, 1]</td>\n",
       "      <td>Canx [1, 2, 1]</td>\n",
       "      <td>ENSMUST00000179865.6</td>\n",
       "      <td>Canx-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50294466</td>\n",
       "      <td>50325673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1</td>\n",
       "      <td>CCDS24633.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012410.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [1, 3, 2]</td>\n",
       "      <td>Canx [1, 3, 2]</td>\n",
       "      <td>ENSMUST00000020637.8</td>\n",
       "      <td>Canx-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50296198</td>\n",
       "      <td>50299378</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012412.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [3, 5, 4]</td>\n",
       "      <td>Canx [3, 5, 4]</td>\n",
       "      <td>ENSMUST00000146979.1</td>\n",
       "      <td>Canx-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50297363</td>\n",
       "      <td>50298319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012411.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [4, 4, 5]</td>\n",
       "      <td>Canx [4, 4, 5]</td>\n",
       "      <td>ENSMUST00000153068.1</td>\n",
       "      <td>Canx-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50298837</td>\n",
       "      <td>50300990</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000020368.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>appris_principal_2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000012413.1</td>\n",
       "      <td>ENSMUSG00000020368.14 [2, 1, 3]</td>\n",
       "      <td>Canx [2, 1, 3]</td>\n",
       "      <td>ENSMUST00000155801.1,ENSMUST00000155801.2</td>\n",
       "      <td>Canx-004,Canx-004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome   Source     Feature     Start       End Score Strand Frame  \\\n",
       "0       chr11  ENSEMBL  transcript  50293960  50325673     .      -     .   \n",
       "35      chr11   HAVANA  transcript  50294466  50325673     .      -     .   \n",
       "70      chr11   HAVANA  transcript  50296198  50299378     .      -     .   \n",
       "76      chr11   HAVANA  transcript  50297363  50298319     .      -     .   \n",
       "79      chr11   HAVANA  transcript  50298837  50300990     .      -     .   \n",
       "\n",
       "                  gene_id       gene_type  ...                 tag  \\\n",
       "0   ENSMUSG00000020368.14  protein_coding  ...                CCDS   \n",
       "35  ENSMUSG00000020368.14  protein_coding  ...                CCDS   \n",
       "70  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "76  ENSMUSG00000020368.14  protein_coding  ...                 NaN   \n",
       "79  ENSMUSG00000020368.14  protein_coding  ...  appris_principal_2   \n",
       "\n",
       "   transcript_support_level       ccdsid exon_number exon_id  \\\n",
       "0                         1  CCDS24633.1         NaN     NaN   \n",
       "35                        1  CCDS24633.1         NaN     NaN   \n",
       "70                        2          NaN         NaN     NaN   \n",
       "76                        5          NaN         NaN     NaN   \n",
       "79                        3          NaN         NaN     NaN   \n",
       "\n",
       "       havana_transcript                    transcript_id transcript_name  \\\n",
       "0                    NaN  ENSMUSG00000020368.14 [1, 2, 1]  Canx [1, 2, 1]   \n",
       "35  OTTMUST00000012410.1  ENSMUSG00000020368.14 [1, 3, 2]  Canx [1, 3, 2]   \n",
       "70  OTTMUST00000012412.1  ENSMUSG00000020368.14 [3, 5, 4]  Canx [3, 5, 4]   \n",
       "76  OTTMUST00000012411.1  ENSMUSG00000020368.14 [4, 4, 5]  Canx [4, 4, 5]   \n",
       "79  OTTMUST00000012413.1  ENSMUSG00000020368.14 [2, 1, 3]  Canx [2, 1, 3]   \n",
       "\n",
       "                       original_transcript_id original_transcript_name  \n",
       "0                        ENSMUST00000179865.6                 Canx-201  \n",
       "35                       ENSMUST00000020637.8                 Canx-001  \n",
       "70                       ENSMUST00000146979.1                 Canx-003  \n",
       "76                       ENSMUST00000153068.1                 Canx-002  \n",
       "79  ENSMUST00000155801.1,ENSMUST00000155801.2        Canx-004,Canx-004  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 622,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Feature == 'transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2bc301b-6c62-4e58-bc60-8e3208718e3f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
